<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2020-02-20" />

<title>Practical – k-mer distributions</title>

<script src="kmer_distributions_solutions_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="kmer_distributions_solutions_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="kmer_distributions_solutions_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="kmer_distributions_solutions_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="kmer_distributions_solutions_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="kmer_distributions_solutions_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="kmer_distributions_solutions_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="kmer_distributions_solutions_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="kmer_distributions_solutions_files/navigation-1.1/tabsets.js"></script>
<script src="kmer_distributions_solutions_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Practical – k-mer distributions</h1>
<h3 class="subtitle">Probabilities and statistics for modelling 1 (STAT1)</h3>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">2020-02-20</h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this practical, we will count k-mer occurrences in DNA sequences of different organisms (one organism per sutdent), fit different theoretical distributions of probabilities onto the empirical distributions of counts, and test the goodness of fit for these alternative distributions.</p>
</div>
<div id="organisms" class="section level2">
<h2>Organisms</h2>
<p>Each student will choose one organism of interest among the following ones.</p>
<table>
<thead>
<tr class="header">
<th>Taxon</th>
<th>Organism</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Fungi</td>
<td>Saccharomyces_cerevisiae</td>
</tr>
<tr class="even">
<td>Bacteria</td>
<td>Escherichia coli GCF 000005845.2 ASM584v2</td>
</tr>
<tr class="odd">
<td>Mammalian</td>
<td>Homo sapiens GRCh38</td>
</tr>
<tr class="even">
<td>Mammalian</td>
<td>Mus musculus GRCm38</td>
</tr>
<tr class="odd">
<td>Bird</td>
<td>Gallus gallus EnsEMBL</td>
</tr>
<tr class="even">
<td>Fish</td>
<td>Danio_rerio_EnsEMBL</td>
</tr>
<tr class="odd">
<td>Insect</td>
<td>Drosophila melanogaster</td>
</tr>
<tr class="even">
<td>Worm</td>
<td></td>
</tr>
<tr class="odd">
<td>Plant</td>
<td>Arabidopsis thaliana.TAIR10.29</td>
</tr>
<tr class="even">
<td>Plant</td>
<td>Zea mays.AGPv3.29</td>
</tr>
<tr class="odd">
<td>Apicomplexa</td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="solutions" class="section level2">
<h2>Solutions</h2>
<div id="working-directory" class="section level3">
<h3>Working directory</h3>
<p>On your computer, create a directory for this practical. I suggest to use a consistent naming for the different practicals of this course.</p>
<p>We will further create a sub-folder with the name of our organism of interest.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">## Define your organism of interest</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">org &lt;-<span class="st"> &quot;Homo_sapiens&quot;</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">## Define a working directory</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">work.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="st">&quot;~&quot;</span>, <span class="st">&quot;CMB-STAT2_practicals&quot;</span>, <span class="st">&quot;kmer_distrib&quot;</span>, org)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="kw">dir.create</span>(work.dir, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">## Print a message with the result directory</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="kw">message</span>(<span class="st">&quot;Result directory</span><span class="ch">\t</span><span class="st">&quot;</span>, work.dir)</a></code></pre></div>
</div>
</div>
<div id="counting-k-mer-occurrences-in-each-promoter-of-a-model-organism" class="section level2">
<h2>Counting k-mer occurrences in each promoter of a model organism</h2>
<ol style="list-style-type: decimal">
<li><p>Open a connection to the Regulatory Sequence Analysis Tools (RSAT) teaching server : <a href="http://teaching.rsat.eu/" class="uri">http://teaching.rsat.eu/</a></p></li>
<li><p>In the tool search box, type “retrieve sequence” and click on the corresponding tool.</p></li>
<li><p>In the <a href="http://pedagogix-tagc.univ-mrs.fr/rsat/retrieve-seq_form.cgi"><em>retrieve-sequence</em></a> form,</p>
<ul>
<li>click <em>Mandatory inputs</em>, enter the name your organism of interest, and check the option <em>all genes of this organism</em>;</li>
<li>in <em>Mandatory options</em>, select <em>upstream</em>, and set the sequence limits from -1 to -500</li>
<li>in <em>Advanced options</em>, <em>make sure that this option is <strong>unchecked</strong>: </em>Prevent overlap with neighbour genes (noorf)* <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></li>
<li>click <em>Run analysis</em> and <em>GO</em>.</li>
</ul></li>
</ol>
<p>After a few seconds (or minutes) the result is displayed. Right-click on the sequence file (extension fasta) and open it in a separate tab to check its content.</p>
<ol start="4" style="list-style-type: decimal">
<li><p>Come back to the result page of retrieve-sequence. In the <em>Next Step</em> box below the result, click on the link to <em>oligo-analysis</em>. This will transfer your sequences to the oligo-analysis form.</p>
<ul>
<li>In the <em>Sequence</em> section, inactivate the option <em>purge sequence</em>.</li>
<li>In the <em>Oligmer counting mode</em>, <strong>uncheck</strong> the option <em>prevent overlapping matches</em>.</li>
<li>Select <em>Count on single strand</em>.</li>
<li>For <em>oligomer lengths</em>, select 2 and <strong>uncheck the other lengths</strong>.</li>
<li>In <em>Results</em>, check the option <em>Occurrence table</em>.</li>
<li>Type your email address and select the mail output.</li>
<li>Click <em>GO</em>.</li>
</ul></li>
</ol>
<p>After a few seconds (minutes), the RSAT server should display the result page, with links to the k-mer count table. Copy the URL of the result file.</p>
<div id="download-the-count-table-from-rsat" class="section level3">
<h3>Download the count table from RSAT</h3>
<p>Let us define the name we will give to the local copy of the k-mer count table generated on the RSAT server in the previous steps.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">## Define the path and the name of the local file</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">kmer.file &lt;-<span class="st"> </span><span class="kw">file.path</span>(work.dir, <span class="st">&quot;2nt-ovlp-1str_Homo_sapiens.tab&quot;</span>)</a></code></pre></div>
<p>One solution is to download manually the k-mer count table generated on the RSAT server, move it to the work directory, and rename it <code>2nt-ovlp-1str_Homo_sapiens.tab</code> (to be adapted depending on your organism of interest).</p>
<p>Another possiblity is to use R command <code>download.file()</code> download it from the URL of the result file on the RSAT server.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">## Note: this will only work for a few days, because the temporary files are removed from the server</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">temp.url &lt;-<span class="st"> &quot;http://pedagogix-tagc.univ-mrs.fr/rsat/tmp/www-data/2020/02/17/oligo-analysis_2020-02-17.155035_zZ4IA4&quot;</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">## Provide the arguments in the order of the function definition</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="kw">download.file</span>(temp.url, kmer.file)</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">## Equivalent : name the arguments</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co"># download.file(url = temp.url, destfile = kmer.file)</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">## Note: named arguments can be provided in a different order without problem</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="co"># download.file(destfile = kmer.file, url = temp.url)</span></a></code></pre></div>
<p>Whichever method was chosen, check that the file is at the right place on your computed.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">## List the files in the working directory</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw">list.files</span>(work.dir)</a></code></pre></div>
<pre><code>[1] &quot;2nt-ovlp-1str_Homo_sapiens.tab&quot;</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co">## Send a message with the k-mer file location</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;K-mer count table file:</span><span class="ch">\t</span><span class="st">&quot;</span>, kmer.file)</a></code></pre></div>
</div>
<div id="load-the-k-mer-count-table-in-r" class="section level3">
<h3>Load the k-mer count table in R</h3>
<p>Use the finction <code>read.table()</code> to load the k-mer count table in a variable named <code>kmer.table</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co">## Call the help for read.table()</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co"># ?read.table</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">## Load the k-mer count table in a variable</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">kmer.table &lt;-<span class="st"> </span><span class="kw">read.table</span>(</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">  <span class="dt">file =</span> kmer.file, </a>
<a class="sourceLine" id="cb7-7" data-line-number="7">  <span class="dt">comment.char =</span> <span class="st">&quot;;&quot;</span>, <span class="co">## comment lines start with &quot;;&quot; in RSAT</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8">  <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="co"># the first row (after the comments) contain the column headers</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9">  <span class="dt">row.names =</span> <span class="dv">1</span>, <span class="co">## the first column contains row names but there might be homonyms</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10">  <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span> <span class="co">## column separator is the tabulation</span></a>
<a class="sourceLine" id="cb7-11" data-line-number="11">  )</a>
<a class="sourceLine" id="cb7-12" data-line-number="12"></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="co">## We set uppercases for the dinucleotide sequences, which were in lowercases in the original count table</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="kw">names</span>(kmer.table) &lt;-<span class="st"> </span><span class="kw">toupper</span>(<span class="kw">names</span>(kmer.table))</a></code></pre></div>
<p>Check the dimensions of this table.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co">## Check the dimensions of the k-mer count table</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw">dim</span>(kmer.table)</a></code></pre></div>
<pre><code>[1] 60675    16</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co">## Number of k-mers</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">m &lt;-<span class="st"> </span><span class="kw">ncol</span>(kmer.table)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="co">## Number of genes</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5">n &lt;-<span class="st"> </span><span class="kw">nrow</span>(kmer.table)</a>
<a class="sourceLine" id="cb10-6" data-line-number="6"></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="co">## Print the result</span></a>
<a class="sourceLine" id="cb10-8" data-line-number="8"><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;Number of rows: &quot;</span>, n))</a></code></pre></div>
<pre><code>[1] &quot;Number of rows: 60675&quot;</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;Number of columns: &quot;</span>, m))</a></code></pre></div>
<pre><code>[1] &quot;Number of columns: 16&quot;</code></pre>
<p>Check the column names</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co">## Check the column names</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="kw">names</span>(kmer.table)</a></code></pre></div>
<pre><code> [1] &quot;AA&quot; &quot;AC&quot; &quot;AG&quot; &quot;AT&quot; &quot;CA&quot; &quot;CC&quot; &quot;CG&quot; &quot;CT&quot; &quot;GA&quot; &quot;GC&quot; &quot;GG&quot; &quot;GT&quot; &quot;TA&quot; &quot;TC&quot; &quot;TG&quot; &quot;TT&quot;</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw">colnames</span>(kmer.table) <span class="co"># equivalent</span></a></code></pre></div>
<pre><code> [1] &quot;AA&quot; &quot;AC&quot; &quot;AG&quot; &quot;AT&quot; &quot;CA&quot; &quot;CC&quot; &quot;CG&quot; &quot;CT&quot; &quot;GA&quot; &quot;GC&quot; &quot;GG&quot; &quot;GT&quot; &quot;TA&quot; &quot;TC&quot; &quot;TG&quot; &quot;TT&quot;</code></pre>
<p>Display the first and last 10 lines of the k-mer count table.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="co">## Show the first 10 lines of the k-mer count table</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="kw">head</span>(kmer.table)</a></code></pre></div>
<pre><code>                                            AA AC AG AT CA CC CG CT GA GC GG GT TA TC TG TT
ENSG00000210049|Homo_sapiens_GRCh38|MT-TF   48 54 17 36 58 69 14 32 14 24  7 11 35 26 18 36
ENSG00000211459|Homo_sapiens_GRCh38|MT-RNR1 56 58 16 36 62 69  9 31 10 19  7 11 38 25 15 37
ENSG00000210077|Homo_sapiens_GRCh38|MT-TV   57 42 41 22 37 38 18 35 28 26 22 27 39 23 22 22
ENSG00000210082|Homo_sapiens_GRCh38|MT-RNR2 48 42 42 22 35 39 17 36 30 25 22 28 41 21 24 27
ENSG00000209082|Homo_sapiens_GRCh38|MT-TL1  52 41 32 34 40 40 19 29 28 16 23 23 38 32 16 34
ENSG00000198888|Homo_sapiens_GRCh38|MT-ND1  48 38 35 32 39 37 21 29 30 16 24 25 36 37 15 35</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw">tail</span>(kmer.table)</a></code></pre></div>
<pre><code>                                                   AA AC AG AT CA CC CG CT GA GC GG GT TA TC TG TT
ENSG00000128973|Homo_sapiens_GRCh38|CLN6           55 24 34 31 31 31 15 36 30 34 29 19 27 24 34 45
ENSG00000272269|Homo_sapiens_GRCh38|RP11-500C11.3  41 29 33 14 43 90 25 33 21 38 32 15 11 35 16 23
ENSG00000267091|Homo_sapiens_GRCh38|CTBP2P7        21 22 38 25 40 34  9 37 30 35 28 32 15 29 50 54
ENSG00000151655|Homo_sapiens_GRCh38|ITIH2          62 30 35 33 48 23  2 41 27 22 17 19 23 39 30 48
ENSG00000234159|Homo_sapiens_GRCh38|RBPMSLP       112 30 39 36 41 34  4 24 32 20 18 14 33 19 23 20
ENSG00000141338|Homo_sapiens_GRCh38|ABCA8          74 22 39 48 31 13  4 30 43 21 19 16 35 22 37 45</code></pre>
</div>
</div>
<div id="compute-marginal-statistics" class="section level2">
<h2>Compute marginal statistics</h2>
<div id="statistics-per-column-k-mer" class="section level3">
<h3>Statistics per column (k-mer)</h3>
<p>Tips: use the R function <code>apply()</code>.</p>
<p>We will compute marginal statistics on the rows and columns of the count table. One possibility is to add columns and rows on this table, but this would not be very convenient for the subsequent computations to be performed on the counts. We will thus create two separate tables: one with the row-wise statistics, and another one with the column-wise statistics.</p>
<p>Rather than looping over each row or column of the count table, we can use the function <code>apply()</code>in order to compute a chosen statistics on each row (margin = 1) or column (margin = 2) of the table.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co">## Recall the dimensions of the kmer count table</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="kw">dim</span>(kmer.table)</a></code></pre></div>
<pre><code>[1] 60675    16</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="co"># View(kmer.table)</span></a>
<a class="sourceLine" id="cb24-2" data-line-number="2"></a>
<a class="sourceLine" id="cb24-3" data-line-number="3"><span class="co">## Compute means per column (on the &quot;second&quot; margin)</span></a>
<a class="sourceLine" id="cb24-4" data-line-number="4">col.means &lt;-<span class="st"> </span><span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, mean)</a>
<a class="sourceLine" id="cb24-5" data-line-number="5"></a>
<a class="sourceLine" id="cb24-6" data-line-number="6"><span class="kw">print</span>(col.means)</a></code></pre></div>
<pre><code>      AA       AC       AG       AT       CA       CC       CG       CT       GA       GC       GG       GT       TA       TC       TG       TT 
40.51309 24.69333 36.14415 28.82729 35.02548 36.94124 14.57330 35.30629 30.15806 30.75028 36.17958 24.01994 24.42770 29.55186 34.20470 37.63466 </code></pre>
<p>We can run the <code>apply()</code>function in the same way to compute the other descriptive statistics (median, variance, …) and store each result in a separate vector. However, it would be more convenient to dispose of a single structure containing all the statistics for each column. For this, we use <code>data.frame()</code> to create a data frame with one column per column per computed statistics, and one row per k-mer.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="co">## Compute a table summarising the statistics per k-mer </span></a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="co">## (marginal stats on the columns)</span></a>
<a class="sourceLine" id="cb26-3" data-line-number="3">stats.per.kmer &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb26-4" data-line-number="4">  <span class="dt">mean =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, mean),</a>
<a class="sourceLine" id="cb26-5" data-line-number="5">  <span class="dt">sd =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, sd),</a>
<a class="sourceLine" id="cb26-6" data-line-number="6">  <span class="dt">var =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, var),</a>
<a class="sourceLine" id="cb26-7" data-line-number="7">  <span class="dt">min =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, min),</a>
<a class="sourceLine" id="cb26-8" data-line-number="8">  <span class="dt">p05 =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, quantile, <span class="fl">0.05</span>),</a>
<a class="sourceLine" id="cb26-9" data-line-number="9">  <span class="dt">Q1 =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, quantile, <span class="fl">0.25</span>),</a>
<a class="sourceLine" id="cb26-10" data-line-number="10">  <span class="dt">median =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, median),</a>
<a class="sourceLine" id="cb26-11" data-line-number="11">  <span class="dt">Q3 =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, quantile, <span class="fl">0.75</span>),</a>
<a class="sourceLine" id="cb26-12" data-line-number="12">  <span class="dt">p95 =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, quantile, <span class="fl">0.95</span>),</a>
<a class="sourceLine" id="cb26-13" data-line-number="13">  <span class="dt">max =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, max),</a>
<a class="sourceLine" id="cb26-14" data-line-number="14">  <span class="dt">sum =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">2</span>, sum)</a>
<a class="sourceLine" id="cb26-15" data-line-number="15">)</a>
<a class="sourceLine" id="cb26-16" data-line-number="16"></a>
<a class="sourceLine" id="cb26-17" data-line-number="17"></a>
<a class="sourceLine" id="cb26-18" data-line-number="18"><span class="kw">kable</span>(stats.per.kmer, <span class="dt">caption =</span> <span class="st">&quot;Column-wise means of the k-mer count table. &quot;</span>)</a></code></pre></div>
<table>
<caption>Column-wise means of the k-mer count table.</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">mean</th>
<th align="right">sd</th>
<th align="right">var</th>
<th align="right">min</th>
<th align="right">p05</th>
<th align="right">Q1</th>
<th align="right">median</th>
<th align="right">Q3</th>
<th align="right">p95</th>
<th align="right">max</th>
<th align="right">sum</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>AA</td>
<td align="right">40.51309</td>
<td align="right">20.706267</td>
<td align="right">428.74949</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="right">25</td>
<td align="right">39</td>
<td align="right">54</td>
<td align="right">77</td>
<td align="right">186</td>
<td align="right">2458132</td>
</tr>
<tr class="even">
<td>AC</td>
<td align="right">24.69333</td>
<td align="right">6.860027</td>
<td align="right">47.05997</td>
<td align="right">0</td>
<td align="right">14</td>
<td align="right">20</td>
<td align="right">24</td>
<td align="right">29</td>
<td align="right">36</td>
<td align="right">161</td>
<td align="right">1498268</td>
</tr>
<tr class="odd">
<td>AG</td>
<td align="right">36.14415</td>
<td align="right">9.103262</td>
<td align="right">82.86939</td>
<td align="right">0</td>
<td align="right">23</td>
<td align="right">30</td>
<td align="right">36</td>
<td align="right">42</td>
<td align="right">51</td>
<td align="right">139</td>
<td align="right">2193046</td>
</tr>
<tr class="even">
<td>AT</td>
<td align="right">28.82729</td>
<td align="right">14.821645</td>
<td align="right">219.68117</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">17</td>
<td align="right">28</td>
<td align="right">39</td>
<td align="right">54</td>
<td align="right">186</td>
<td align="right">1749096</td>
</tr>
<tr class="odd">
<td>CA</td>
<td align="right">35.02548</td>
<td align="right">8.344418</td>
<td align="right">69.62931</td>
<td align="right">0</td>
<td align="right">22</td>
<td align="right">30</td>
<td align="right">35</td>
<td align="right">40</td>
<td align="right">48</td>
<td align="right">173</td>
<td align="right">2125171</td>
</tr>
<tr class="even">
<td>CC</td>
<td align="right">36.94124</td>
<td align="right">20.350163</td>
<td align="right">414.12914</td>
<td align="right">0</td>
<td align="right">12</td>
<td align="right">22</td>
<td align="right">33</td>
<td align="right">47</td>
<td align="right">76</td>
<td align="right">208</td>
<td align="right">2241410</td>
</tr>
<tr class="odd">
<td>CG</td>
<td align="right">14.57330</td>
<td align="right">16.515726</td>
<td align="right">272.76920</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">3</td>
<td align="right">8</td>
<td align="right">20</td>
<td align="right">51</td>
<td align="right">113</td>
<td align="right">884235</td>
</tr>
<tr class="even">
<td>CT</td>
<td align="right">35.30629</td>
<td align="right">8.887525</td>
<td align="right">78.98810</td>
<td align="right">0</td>
<td align="right">22</td>
<td align="right">29</td>
<td align="right">35</td>
<td align="right">41</td>
<td align="right">50</td>
<td align="right">137</td>
<td align="right">2142209</td>
</tr>
<tr class="odd">
<td>GA</td>
<td align="right">30.15806</td>
<td align="right">8.455523</td>
<td align="right">71.49587</td>
<td align="right">0</td>
<td align="right">18</td>
<td align="right">25</td>
<td align="right">30</td>
<td align="right">35</td>
<td align="right">44</td>
<td align="right">139</td>
<td align="right">1829840</td>
</tr>
<tr class="even">
<td>GC</td>
<td align="right">30.75028</td>
<td align="right">15.779050</td>
<td align="right">248.97841</td>
<td align="right">0</td>
<td align="right">12</td>
<td align="right">19</td>
<td align="right">27</td>
<td align="right">38</td>
<td align="right">63</td>
<td align="right">113</td>
<td align="right">1865773</td>
</tr>
<tr class="odd">
<td>GG</td>
<td align="right">36.17958</td>
<td align="right">19.283920</td>
<td align="right">371.86955</td>
<td align="right">0</td>
<td align="right">12</td>
<td align="right">22</td>
<td align="right">33</td>
<td align="right">46</td>
<td align="right">73</td>
<td align="right">208</td>
<td align="right">2195196</td>
</tr>
<tr class="even">
<td>GT</td>
<td align="right">24.01994</td>
<td align="right">6.850075</td>
<td align="right">46.92352</td>
<td align="right">0</td>
<td align="right">14</td>
<td align="right">20</td>
<td align="right">24</td>
<td align="right">28</td>
<td align="right">35</td>
<td align="right">130</td>
<td align="right">1457410</td>
</tr>
<tr class="odd">
<td>TA</td>
<td align="right">24.42770</td>
<td align="right">13.962177</td>
<td align="right">194.94237</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">13</td>
<td align="right">23</td>
<td align="right">34</td>
<td align="right">49</td>
<td align="right">184</td>
<td align="right">1482151</td>
</tr>
<tr class="even">
<td>TC</td>
<td align="right">29.55186</td>
<td align="right">8.177379</td>
<td align="right">66.86953</td>
<td align="right">0</td>
<td align="right">17</td>
<td align="right">24</td>
<td align="right">29</td>
<td align="right">34</td>
<td align="right">43</td>
<td align="right">139</td>
<td align="right">1793059</td>
</tr>
<tr class="odd">
<td>TG</td>
<td align="right">34.20470</td>
<td align="right">8.940656</td>
<td align="right">79.93532</td>
<td align="right">0</td>
<td align="right">20</td>
<td align="right">28</td>
<td align="right">34</td>
<td align="right">40</td>
<td align="right">48</td>
<td align="right">144</td>
<td align="right">2075370</td>
</tr>
<tr class="even">
<td>TT</td>
<td align="right">37.63466</td>
<td align="right">19.433749</td>
<td align="right">377.67061</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="right">23</td>
<td align="right">36</td>
<td align="right">50</td>
<td align="right">72</td>
<td align="right">163</td>
<td align="right">2283483</td>
</tr>
</tbody>
</table>
</div>
<div id="warning-about-r-sd-and-var-functions" class="section level3">
<h3>Warning about R sd() and var() functions</h3>
<p><strong>Beware:</strong> the R functions <code>sd</code> and <code>var</code> do not compute the standard deviation and variance of the numbers provded! Instead, they consider that the data provided are a sample drawed from a population, and the return an estimate of the standard deviation or variance of this population.</p>
<p>So, instead of the sample variance</p>
<p><span class="math display">\[s^2 = \sum_{i=1}^n \frac{1}{n}(x - \bar{x})^2\]</span></p>
<p>They estimate the variance of the population corrected for the systematic bias of sample variance.</p>
<p><span class="math display">\[\hat{s^2} = \sum_{i=1}^n \frac{1}{n-1}(x - \bar{x})^2\]</span></p>
<p>For this exercise, this is actually what we want, since we can consider that the 60675 genes of our organism of interest are a sampling of all the genes that might be found in organisms of the same taxon. In any case, the impact of the correction is negligible with such a large number of genes (<span class="math inline">\(n = 60675\)</span>).</p>
<p><span class="math display">\[\frac{n}{n-1} =  1.0000165\]</span></p>
</div>
<div id="analyse-the-relationships-between-the-mean-and-the-variance" class="section level3">
<h3>Analyse the relationships between the mean and the variance</h3>
<p>Poisson: variance = mean Binomial: variance &lt; mean XXX: variance &gt; mean</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">plot</span>(stats.per.kmer<span class="op">$</span>mean, stats.per.kmer<span class="op">$</span>var, </a>
<a class="sourceLine" id="cb27-2" data-line-number="2">     <span class="dt">xlab =</span> <span class="st">&quot;Mean of the counts per promoters&quot;</span>, </a>
<a class="sourceLine" id="cb27-3" data-line-number="3">     <span class="dt">ylab =</span> <span class="st">&quot;Variance of the counts per promoters&quot;</span>, </a>
<a class="sourceLine" id="cb27-4" data-line-number="4">     <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">panel.first =</span> <span class="kw">grid</span>(),</a>
<a class="sourceLine" id="cb27-5" data-line-number="5">     <span class="dt">main =</span> <span class="st">&quot;Dinucleotide counts in human promoters</span><span class="ch">\n</span><span class="st">Variance versus mean plot&quot;</span>,</a>
<a class="sourceLine" id="cb27-6" data-line-number="6">     <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">10</span>,<span class="dv">50</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">500</span>), <span class="dt">cex =</span> <span class="dv">0</span>)</a>
<a class="sourceLine" id="cb27-7" data-line-number="7"><span class="kw">text</span>(<span class="dt">x =</span> stats.per.kmer<span class="op">$</span>mean, <span class="dt">y =</span> stats.per.kmer<span class="op">$</span>var, <span class="dt">labels =</span> <span class="kw">rownames</span>(stats.per.kmer))</a>
<a class="sourceLine" id="cb27-8" data-line-number="8"><span class="kw">abline</span>(<span class="dt">a =</span> <span class="dv">0</span>, <span class="dt">b =</span> <span class="dv">1</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_mean_vs_var-1.png" alt="Mean-variance plot for dinucleotide counts in all human promoters. The line indicates the variance/mean relationship in Poisson. distributions. " width="60%" />
<p class="caption">
Mean-variance plot for dinucleotide counts in all human promoters. The line indicates the variance/mean relationship in Poisson. distributions.
</p>
</div>
</div>
<div id="statistics-per-gene-row" class="section level3">
<h3>Statistics per gene (row)</h3>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="co">## Compute a table summarising the gene-wise (row-wise) statistics of the k-mer count table</span></a>
<a class="sourceLine" id="cb28-2" data-line-number="2">stats.per.gene &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb28-3" data-line-number="3">  <span class="dt">mean =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, mean),</a>
<a class="sourceLine" id="cb28-4" data-line-number="4">  <span class="dt">sd =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, sd),</a>
<a class="sourceLine" id="cb28-5" data-line-number="5">  <span class="dt">var =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, var),</a>
<a class="sourceLine" id="cb28-6" data-line-number="6">  <span class="dt">min =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, min),</a>
<a class="sourceLine" id="cb28-7" data-line-number="7">  <span class="dt">Q1 =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, quantile, <span class="fl">0.25</span>),</a>
<a class="sourceLine" id="cb28-8" data-line-number="8">  <span class="dt">median =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, median),</a>
<a class="sourceLine" id="cb28-9" data-line-number="9">  <span class="dt">Q3 =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, quantile, <span class="fl">0.75</span>),</a>
<a class="sourceLine" id="cb28-10" data-line-number="10">  <span class="dt">max =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, max),</a>
<a class="sourceLine" id="cb28-11" data-line-number="11">  <span class="dt">sum =</span> <span class="kw">apply</span>(kmer.table, <span class="dv">1</span>, sum)</a>
<a class="sourceLine" id="cb28-12" data-line-number="12">)</a></code></pre></div>
<p>We can inspect the sum of counts per gene.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw">kable</span>(<span class="kw">rev</span>(<span class="kw">table</span>(stats.per.gene<span class="op">$</span>sum)))</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">499</td>
<td align="right">60658</td>
</tr>
<tr class="even">
<td align="left">497</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">482</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">476</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">465</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">456</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">398</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">393</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">249</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">194</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">142</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">138</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">102</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">24</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">0</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
<p>For almost all genes, the sum of k-mer counts is <span class="math inline">\(n = L -1\)</span>, where <span class="math inline">\(L\)</span> is the sequence length. This is logical since there is no dinucleotide starting at the last position of a sequence. More generally, for a k-mer of size <span class="math inline">\(k\)</span>, the number of possible positions is <span class="math inline">\(n = L - k + 1\)</span>.</p>
<p>The very few cases of sequences having a smaller sums of counts result from the presence of ambiguous nucleotides (represented by the symbol <code>N</code>) in the genome sequence. For this exercise, we will suppress these genes because our models rely on the hypothesis that the sum of counts per gene is constant.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="co">## Filter out the genes having an incorrect sum of counts</span></a>
<a class="sourceLine" id="cb30-2" data-line-number="2">n &lt;-<span class="st"> </span><span class="kw">max</span>(stats.per.gene<span class="op">$</span>sum)</a>
<a class="sourceLine" id="cb30-3" data-line-number="3"></a>
<a class="sourceLine" id="cb30-4" data-line-number="4"><span class="co">## Count the number of genes with the correct sum of counts</span></a>
<a class="sourceLine" id="cb30-5" data-line-number="5"><span class="kw">sum</span>(stats.per.gene<span class="op">$</span>sum <span class="op">==</span><span class="st"> </span>n)</a></code></pre></div>
<pre><code>[1] 60658</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="co">## Count the number of genes with an incorrect sum of counts</span></a>
<a class="sourceLine" id="cb32-2" data-line-number="2"><span class="kw">sum</span>(stats.per.gene<span class="op">$</span>sum <span class="op">!=</span><span class="st"> </span>n)</a></code></pre></div>
<pre><code>[1] 17</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="co">## Filter out the genes with incorrect sums of counts</span></a>
<a class="sourceLine" id="cb34-2" data-line-number="2">kmer.table &lt;-<span class="st"> </span>kmer.table[stats.per.gene<span class="op">$</span>sum <span class="op">==</span><span class="st"> </span>n, ]</a>
<a class="sourceLine" id="cb34-3" data-line-number="3"><span class="kw">dim</span>(kmer.table)</a></code></pre></div>
<pre><code>[1] 60658    16</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="co">## Count the remaining number of genes</span></a>
<a class="sourceLine" id="cb36-2" data-line-number="2">nb.genes &lt;-<span class="st"> </span><span class="kw">nrow</span>(kmer.table)</a></code></pre></div>
</div>
</div>
<div id="drawing-empirical-distributions-of-counts" class="section level2">
<h2>Drawing empirical distributions of counts</h2>
<div id="draw-an-histogram-with-the-distribution-of-counts-for-a-given-k-mer." class="section level3">
<h3>Draw an histogram with the distribution of counts for a given k-mer.</h3>
<p>The simplest way to generate a histogram is to use the <code>hist()</code> function. Here is the result with the default parameters.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1">kmer &lt;-<span class="st"> &quot;AG&quot;</span> <span class="co"># Select a k-mer</span></a>
<a class="sourceLine" id="cb37-2" data-line-number="2">counts &lt;-<span class="st"> </span>kmer.table[, kmer] <span class="co"># select the counts for this k-mer</span></a>
<a class="sourceLine" id="cb37-3" data-line-number="3"><span class="kw">range</span>(counts)</a></code></pre></div>
<pre><code>[1]   0 139</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="co">## Draw an histogram with default parameters (not very beautiful)</span></a>
<a class="sourceLine" id="cb39-2" data-line-number="2"><span class="kw">hist</span>(<span class="dt">x =</span> counts)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_one_kmer_hist-1.png" alt="Counts per sequence in all human promoters. " width="60%" />
<p class="caption">
Counts per sequence in all human promoters.
</p>
</div>
<p>We can fine-tune the parameters in order to get a more informative representation, by specifying custom breaks.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="co">## Draw a nice and informative histogram</span></a>
<a class="sourceLine" id="cb40-2" data-line-number="2"><span class="kw">hist</span>(<span class="dt">x =</span> counts, </a>
<a class="sourceLine" id="cb40-3" data-line-number="3">     <span class="dt">breaks =</span> <span class="dv">0</span><span class="op">:</span><span class="kw">max</span>(counts <span class="op">+</span><span class="st"> </span><span class="dv">1</span>),</a>
<a class="sourceLine" id="cb40-4" data-line-number="4">     <span class="dt">col =</span> <span class="st">&quot;#44DDFF&quot;</span>, <span class="dt">border =</span> <span class="st">&quot;#44DDFF&quot;</span>,</a>
<a class="sourceLine" id="cb40-5" data-line-number="5">     <span class="dt">las =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb40-6" data-line-number="6">     <span class="dt">ylab =</span> <span class="st">&quot;Number of promoters&quot;</span>,   </a>
<a class="sourceLine" id="cb40-7" data-line-number="7">     <span class="dt">xlab =</span> <span class="st">&quot;K-mer counts&quot;</span>,</a>
<a class="sourceLine" id="cb40-8" data-line-number="8">     <span class="dt">main =</span> <span class="kw">paste0</span>(kmer, <span class="st">&quot; counts in promoters&quot;</span>))</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_one_kmer_hist_pretty-1.png" alt="Counts per sequence in all human promoters. " width="60%" />
<p class="caption">
Counts per sequence in all human promoters.
</p>
</div>
</div>
<div id="generate-a-figure-with-4-x-4-panels-to-depict-the-histograms-of-the-16-k-mers" class="section level3">
<h3>Generate a figure with 4 x 4 panels to depict the histograms of the 16 k-mers</h3>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>))</a>
<a class="sourceLine" id="cb41-2" data-line-number="2"><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="fl">4.1</span>, <span class="fl">5.1</span>, <span class="fl">4.1</span>, <span class="fl">1.1</span>))</a>
<a class="sourceLine" id="cb41-3" data-line-number="3"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">16</span>) {</a>
<a class="sourceLine" id="cb41-4" data-line-number="4">  kmer &lt;-<span class="st"> </span><span class="kw">names</span>(kmer.table)[i]</a>
<a class="sourceLine" id="cb41-5" data-line-number="5">  counts &lt;-<span class="st"> </span>kmer.table[, i] <span class="co"># select the ith column of the k-mer table</span></a>
<a class="sourceLine" id="cb41-6" data-line-number="6">  <span class="co">## Draw a nice and informative histogram</span></a>
<a class="sourceLine" id="cb41-7" data-line-number="7">  <span class="kw">hist</span>(<span class="dt">x =</span> counts, </a>
<a class="sourceLine" id="cb41-8" data-line-number="8">       <span class="dt">breaks =</span> <span class="dv">0</span><span class="op">:</span><span class="kw">max</span>(counts <span class="op">+</span><span class="st"> </span><span class="dv">1</span>),</a>
<a class="sourceLine" id="cb41-9" data-line-number="9">       <span class="dt">col =</span> <span class="st">&quot;#44DDFF&quot;</span>, <span class="dt">border =</span> <span class="st">&quot;#44DDFF&quot;</span>,</a>
<a class="sourceLine" id="cb41-10" data-line-number="10">       <span class="dt">las =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb41-11" data-line-number="11">       <span class="dt">ylab =</span> <span class="st">&quot;Number of promoters&quot;</span>,   </a>
<a class="sourceLine" id="cb41-12" data-line-number="12">       <span class="dt">xlab =</span> <span class="st">&quot;K-mer counts&quot;</span>,</a>
<a class="sourceLine" id="cb41-13" data-line-number="13">       <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="kw">toupper</span>(kmer), <span class="st">&quot; counts in promoters&quot;</span>))</a>
<a class="sourceLine" id="cb41-14" data-line-number="14">}</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_all_kmer_hist_pretty-1.png" alt="Dinucleotide counts per sequence in all human promoters. " width="100%" />
<p class="caption">
Dinucleotide counts per sequence in all human promoters.
</p>
</div>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</a>
<a class="sourceLine" id="cb42-2" data-line-number="2"><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="fl">4.1</span>, <span class="fl">5.1</span>, <span class="fl">4.1</span>, <span class="fl">2.1</span>))</a></code></pre></div>
</div>
<div id="use-other-graphical-representations-to-get-an-insight-of-the-k-mer-count-distributions-boxplots-violin-plots" class="section level3">
<h3>Use other graphical representations to get an insight of the k-mer count distributions (boxplots, violin plots)</h3>
<div id="boxplot" class="section level4">
<h4>Boxplot</h4>
<p>The funciton <code>boxplot()</code> provides an informative summary of the data.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1"><span class="kw">boxplot</span>(<span class="dt">x =</span> kmer.table)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_kmer_count_boxplot-1.png" alt="Boxplot of dinucleotide counts in all promoters. " width="60%" />
<p class="caption">
Boxplot of dinucleotide counts in all promoters.
</p>
</div>
<p>We can fine-tue the parameters to enhance the readability</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="kw">boxplot</span>(<span class="dt">x =</span> kmer.table[,<span class="kw">order</span>(stats.per.kmer<span class="op">$</span>median)], </a>
<a class="sourceLine" id="cb44-2" data-line-number="2">        <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb44-3" data-line-number="3">        <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb44-4" data-line-number="4">        <span class="dt">col =</span> <span class="st">&quot;#44DDFF&quot;</span>, </a>
<a class="sourceLine" id="cb44-5" data-line-number="5">        <span class="dt">main =</span> <span class="st">&quot;Dinucleotide distributions&quot;</span>, </a>
<a class="sourceLine" id="cb44-6" data-line-number="6">        <span class="dt">xlab =</span> <span class="st">&quot;Counts&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_kmer_count_boxplot_pretty-1.png" alt="Boxplot of dinucleotide counts in all promoters. " width="60%" />
<p class="caption">
Boxplot of dinucleotide counts in all promoters.
</p>
</div>
<p>Alternative: sort by variance.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1"><span class="kw">boxplot</span>(<span class="dt">x =</span> kmer.table[,<span class="kw">order</span>(stats.per.kmer<span class="op">$</span>var)], </a>
<a class="sourceLine" id="cb45-2" data-line-number="2">        <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb45-3" data-line-number="3">        <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb45-4" data-line-number="4">        <span class="dt">col =</span> <span class="st">&quot;#44DDFF&quot;</span>, </a>
<a class="sourceLine" id="cb45-5" data-line-number="5">        <span class="dt">main =</span> <span class="st">&quot;Dinucleotide distributions&quot;</span>, </a>
<a class="sourceLine" id="cb45-6" data-line-number="6">        <span class="dt">xlab =</span> <span class="st">&quot;Counts&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_kmer_count_boxplot_pretty_var-1.png" alt="Boxplot of dinucleotide counts in all promoters. " width="60%" />
<p class="caption">
Boxplot of dinucleotide counts in all promoters.
</p>
</div>
</div>
<div id="violin-plot" class="section level4">
<h4>Violin plot</h4>
<p>We can use the R function <code>vioplot()</code> to generate a violin plot. This requires to install the R package <code>vioplot</code>, if it is not already done.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1"><span class="co">## Check if the vioplot package is already installed</span></a>
<a class="sourceLine" id="cb46-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw">require</span>(vioplot)) {</a>
<a class="sourceLine" id="cb46-3" data-line-number="3">  <span class="co">## Install the vioplot package</span></a>
<a class="sourceLine" id="cb46-4" data-line-number="4">  <span class="kw">install.packages</span>(<span class="st">&quot;vioplot&quot;</span>)</a>
<a class="sourceLine" id="cb46-5" data-line-number="5">}</a>
<a class="sourceLine" id="cb46-6" data-line-number="6"><span class="kw">library</span>(vioplot) <span class="co">## Load the vioplot package</span></a>
<a class="sourceLine" id="cb46-7" data-line-number="7"><span class="co"># help(package = vioplot) # Get help on the package</span></a>
<a class="sourceLine" id="cb46-8" data-line-number="8"><span class="co"># help(vioplot) # Get help on the vioplot function</span></a>
<a class="sourceLine" id="cb46-9" data-line-number="9"></a>
<a class="sourceLine" id="cb46-10" data-line-number="10"><span class="kw">vioplot</span>(kmer.table)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_violin_plot-1.png" alt="Violin plot of dinucleotide counts in promoter sequences" width="60%" />
<p class="caption">
Violin plot of dinucleotide counts in promoter sequences
</p>
</div>
<p>Let us fine-tune the parameters to get a pretty violin plot.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1"><span class="kw">vioplot</span>(kmer.table[, <span class="kw">order</span>(stats.per.kmer<span class="op">$</span>mean)], <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, <span class="dt">col =</span> <span class="st">&quot;#44DDFF&quot;</span>, <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">main =</span> <span class="st">&quot;Violin plot of dinucleotide counts in promoters&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;Counts&quot;</span>, )</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_violin_plot_pretty-1.png" alt="Violin plot of dinucleotide counts in promoter sequences" width="60%" />
<p class="caption">
Violin plot of dinucleotide counts in promoter sequences
</p>
</div>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1"><span class="kw">vioplot</span>(kmer.table[, <span class="kw">order</span>(stats.per.kmer<span class="op">$</span>var)], <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, <span class="dt">col =</span> <span class="st">&quot;#44DDFF&quot;</span>, <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">main =</span> <span class="st">&quot;Violin plot of dinucleotide counts in promoters&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;Counts&quot;</span>, )</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_violin_plot_pretty_var-1.png" alt="Violin plot of dinucleotide counts in promoter sequences" width="60%" />
<p class="caption">
Violin plot of dinucleotide counts in promoter sequences
</p>
</div>
</div>
</div>
</div>
<div id="exploring-k-mer-count-distributions-in-promoter-sequences" class="section level2">
<h2>Exploring k-mer count distributions in promoter sequences</h2>
<ol start="5" style="list-style-type: decimal">
<li><p>Compute a vector with the relative frequency of each k-mer in all the sequences.</p></li>
<li><p>Compute a table with the relative frequencies of k-mers per sequence, and compute similar summary statistics per column on this relative frequency table.</p></li>
<li><p>Write a brief interpretation of the results.</p></li>
</ol>
</div>
<div id="fitting" class="section level2">
<h2>Fitting</h2>
<div id="encapsulate-the-pretty-histogram-in-an-r-function" class="section level3">
<h3>Encapsulate the pretty histogram in an R function</h3>
<p>For the rest of this tutorial, we will have to draw several histograms of k-mer count distributions. Since we are satisfied of the fine-tuned histogram above, one possibility is to include it in a function, that we will be able to recall later rather than having to re-type the full code for each usage.</p>
<p>We will properly document this function with the roxygen2 standard. This will enable us to incorporate this function in a custom R package if we wish.</p>
<p>We can also include parameters (defined in the signature of the function) in order to make its use more flexible.</p>
<p>Moreover, this function will return an object (of class “histogram”) with the numerical values of the histogram (breaks, frequencies, counts, …).</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1"><span class="co">#&#39; @title K-mer count histogram</span></a>
<a class="sourceLine" id="cb49-2" data-line-number="2"><span class="co">#&#39; @author Jacques van Helden</span></a>
<a class="sourceLine" id="cb49-3" data-line-number="3"><span class="co">#&#39; @description Draw an histogram with the distribution of a vector of counts for a given k-mer.</span></a>
<a class="sourceLine" id="cb49-4" data-line-number="4"><span class="co">#&#39; @param counts a vector of counts</span></a>
<a class="sourceLine" id="cb49-5" data-line-number="5"><span class="co">#&#39; @param kmer the sequence of the oligonucleotide (k-mer) to be displayed in the title</span></a>
<a class="sourceLine" id="cb49-6" data-line-number="6"><span class="co">#&#39; @param class.interval=1 class interval (breaks are automatically computed)</span></a>
<a class="sourceLine" id="cb49-7" data-line-number="7"><span class="co">#&#39; @param col=&quot;#44DDFF&quot; fill color for the histogram rectangles</span></a>
<a class="sourceLine" id="cb49-8" data-line-number="8"><span class="co">#&#39; @param border=&quot;#22BBFF&quot; border color for the histogram rectangles</span></a>
<a class="sourceLine" id="cb49-9" data-line-number="9"><span class="co">#&#39; @param main=paste0(kmer, &quot; counts in promoters&quot;)</span></a>
<a class="sourceLine" id="cb49-10" data-line-number="10"><span class="co">#&#39; @param ... all the other arguments are passed to the hist function</span></a>
<a class="sourceLine" id="cb49-11" data-line-number="11"><span class="co">#&#39; @return an object of type histogram returned by the hist() function</span></a>
<a class="sourceLine" id="cb49-12" data-line-number="12"><span class="co">#&#39; @export </span></a>
<a class="sourceLine" id="cb49-13" data-line-number="13">kmer.hist &lt;-<span class="st"> </span><span class="cf">function</span>(counts, </a>
<a class="sourceLine" id="cb49-14" data-line-number="14">                      kmer, </a>
<a class="sourceLine" id="cb49-15" data-line-number="15">                      <span class="dt">class.interval =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb49-16" data-line-number="16">                      <span class="dt">col =</span> <span class="st">&quot;#44DDFF&quot;</span>, </a>
<a class="sourceLine" id="cb49-17" data-line-number="17">                      <span class="dt">border =</span> <span class="st">&quot;#44DDFF&quot;</span>,</a>
<a class="sourceLine" id="cb49-18" data-line-number="18">                      <span class="dt">main =</span> <span class="kw">paste0</span>(kmer, <span class="st">&quot; counts in promoters&quot;</span>),</a>
<a class="sourceLine" id="cb49-19" data-line-number="19">                      ...) {</a>
<a class="sourceLine" id="cb49-20" data-line-number="20">  h &lt;-<span class="st"> </span><span class="kw">hist</span>(<span class="dt">x =</span> counts, </a>
<a class="sourceLine" id="cb49-21" data-line-number="21">            <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, </a>
<a class="sourceLine" id="cb49-22" data-line-number="22">                         <span class="dt">to =</span> <span class="kw">max</span>(counts <span class="op">+</span><span class="st"> </span>class.interval), </a>
<a class="sourceLine" id="cb49-23" data-line-number="23">                         <span class="dt">by =</span> class.interval),</a>
<a class="sourceLine" id="cb49-24" data-line-number="24">            <span class="dt">col =</span> col, </a>
<a class="sourceLine" id="cb49-25" data-line-number="25">            <span class="dt">border =</span> border,</a>
<a class="sourceLine" id="cb49-26" data-line-number="26">            <span class="dt">las =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb49-27" data-line-number="27">            <span class="dt">ylab =</span> <span class="st">&quot;Number of promoters&quot;</span>,   </a>
<a class="sourceLine" id="cb49-28" data-line-number="28">            <span class="dt">xlab =</span> <span class="st">&quot;K-mer counts&quot;</span>,</a>
<a class="sourceLine" id="cb49-29" data-line-number="29">            <span class="dt">main =</span> main,</a>
<a class="sourceLine" id="cb49-30" data-line-number="30">            ...)</a>
<a class="sourceLine" id="cb49-31" data-line-number="31">  </a>
<a class="sourceLine" id="cb49-32" data-line-number="32">  <span class="co">## Return the histogram values</span></a>
<a class="sourceLine" id="cb49-33" data-line-number="33">  <span class="kw">return</span>(h)</a>
<a class="sourceLine" id="cb49-34" data-line-number="34">}</a></code></pre></div>
<p>We can then call this function with a given k-mer.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" data-line-number="1">hist.values &lt;-<span class="st"> </span><span class="kw">kmer.hist</span>(<span class="dt">counts =</span> kmer.table[, <span class="st">&quot;TT&quot;</span>], <span class="dt">kmer =</span> <span class="st">&quot;TT&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_another_kmer-1.png" alt="Distribution of k-mer counts in promoter generated with our custom funciton kmer.hist(). " width="60%" />
<p class="caption">
Distribution of k-mer counts in promoter generated with our custom funciton kmer.hist().
</p>
</div>
<p>We can then re-use it with wider breaks.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1">kmer &lt;-<span class="st"> &quot;TT&quot;</span></a>
<a class="sourceLine" id="cb51-2" data-line-number="2">counts &lt;-<span class="st"> </span>kmer.table[, kmer]</a>
<a class="sourceLine" id="cb51-3" data-line-number="3">hist.values.ci5 &lt;-<span class="st"> </span><span class="kw">kmer.hist</span>(<span class="dt">counts =</span> counts, </a>
<a class="sourceLine" id="cb51-4" data-line-number="4">                             <span class="dt">kmer =</span> kmer, </a>
<a class="sourceLine" id="cb51-5" data-line-number="5">                             <span class="dt">border =</span> <span class="st">&quot;#0088FF&quot;</span>,</a>
<a class="sourceLine" id="cb51-6" data-line-number="6">                             <span class="dt">class.interval =</span> <span class="dv">5</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_another_kmer_ci5-1.png" alt="Distribution of k-mer counts in promoter generated with our custom funciton kmer.hist() with class intervals of size 5. " width="60%" />
<p class="caption">
Distribution of k-mer counts in promoter generated with our custom funciton kmer.hist() with class intervals of size 5.
</p>
</div>
<p>We can now extract the values of the histogram, for example the class midpoints and the counts per class.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1"><span class="kw">class</span>(hist.values)</a></code></pre></div>
<pre><code>[1] &quot;histogram&quot;</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1"><span class="kw">names</span>(hist.values)</a></code></pre></div>
<pre><code>[1] &quot;breaks&quot;   &quot;counts&quot;   &quot;density&quot;  &quot;mids&quot;     &quot;xname&quot;    &quot;equidist&quot;</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" data-line-number="1"><span class="co"># View(hist.values)</span></a></code></pre></div>
<p>We can print the midpoints of the class intervals.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" data-line-number="1"><span class="kw">print</span>(hist.values<span class="op">$</span>mids)</a></code></pre></div>
<pre><code>  [1]   0.5   1.5   2.5   3.5   4.5   5.5   6.5   7.5   8.5   9.5  10.5  11.5  12.5  13.5  14.5  15.5  16.5  17.5  18.5  19.5  20.5  21.5  22.5  23.5  24.5  25.5  26.5  27.5  28.5  29.5  30.5  31.5  32.5  33.5  34.5  35.5  36.5  37.5  38.5  39.5  40.5  41.5  42.5  43.5  44.5  45.5  46.5  47.5  48.5
 [50]  49.5  50.5  51.5  52.5  53.5  54.5  55.5  56.5  57.5  58.5  59.5  60.5  61.5  62.5  63.5  64.5  65.5  66.5  67.5  68.5  69.5  70.5  71.5  72.5  73.5  74.5  75.5  76.5  77.5  78.5  79.5  80.5  81.5  82.5  83.5  84.5  85.5  86.5  87.5  88.5  89.5  90.5  91.5  92.5  93.5  94.5  95.5  96.5  97.5
 [99]  98.5  99.5 100.5 101.5 102.5 103.5 104.5 105.5 106.5 107.5 108.5 109.5 110.5 111.5 112.5 113.5 114.5 115.5 116.5 117.5 118.5 119.5 120.5 121.5 122.5 123.5 124.5 125.5 126.5 127.5 128.5 129.5 130.5 131.5 132.5 133.5 134.5 135.5 136.5 137.5 138.5 139.5 140.5 141.5 142.5 143.5 144.5 145.5 146.5
[148] 147.5 148.5 149.5 150.5 151.5 152.5 153.5 154.5 155.5 156.5 157.5 158.5 159.5 160.5 161.5 162.5 163.5</code></pre>
<p>And the corresponding counts.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1"><span class="kw">print</span>(hist.values<span class="op">$</span>counts)</a></code></pre></div>
<pre><code>  [1]   92  100  161  232  298  377  457  481  582  684  721  765  818  874  872  996  988 1031 1043 1092 1097 1156 1113 1150 1145 1128 1203 1170 1205 1182 1221 1120 1182 1133 1145 1204 1187 1126 1130 1143 1074 1067 1097 1037 1052  976  959  955  907  867  887  817  827  720  717  698  635  661  635
 [60]  570  568  519  501  486  406  379  361  343  338  317  293  273  223  225  212  187  166  161  140  132  120  116   99   94  104   85   77   76   50   66   48   51   44   35   30   39   27   29   21   21   10   21   19   23   12   10   14   12   15   10    3   11   13    4    4    9    4    2
[119]    3    2    3    2    1    1    0    1    7    0    1    3    3    0    1    2    2    0    0    1    0    1    0    0    0    0    0    0    0    1    0    0    1    0    0    0    0    0    0    0    0    1    0    0    1    0</code></pre>
<p>We can also use this information to draw a custom plot of the distribution (for example a frequency polygon).</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" data-line-number="1"><span class="kw">plot</span>(hist.values<span class="op">$</span>mids, hist.values<span class="op">$</span>counts, </a>
<a class="sourceLine" id="cb61-2" data-line-number="2">     <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">xlab =</span> <span class="st">&quot;Number of sequences&quot;</span>, </a>
<a class="sourceLine" id="cb61-3" data-line-number="3">     <span class="dt">ylab =</span> <span class="st">&quot;Counts per sequence&quot;</span>,</a>
<a class="sourceLine" id="cb61-4" data-line-number="4">     <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;#008800&quot;</span>, </a>
<a class="sourceLine" id="cb61-5" data-line-number="5">     <span class="dt">lwd =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb61-6" data-line-number="6">     <span class="dt">main =</span> <span class="kw">paste0</span>(kmer, <span class="st">&quot; counts</span><span class="ch">\n</span><span class="st">Frequency polygon&quot;</span>))</a>
<a class="sourceLine" id="cb61-7" data-line-number="7"><span class="kw">grid</span>() <span class="co"># Add a grid to the plot</span></a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_frequency_polygon-1.png" alt="Distribution of counts for the TT dinucleotide in Human promoter sequences. Representation. as frequency polygon. " width="60%" />
<p class="caption">
Distribution of counts for the TT dinucleotide in Human promoter sequences. Representation. as frequency polygon.
</p>
</div>
</div>
<div id="fit-a-poisson-distribution-on-each-empirical-distribution-of-k-mer-counts." class="section level3">
<h3>Fit a Poisson distribution on each empirical distribution of k-mer counts.</h3>
<pre><code>- How do you choose the parameters?
- Draw the fitted Poisson distribution over the histogram of empirical distribution (observed k-mer occurences)

    **Tips:** 
    
    - in order to add some plot over an existing plot, you can use the `lines()` function
    - you can also use specific options to draw histogram-like lines: `lines(x, y, type = &quot;h&quot;). 
    
    </code></pre>
<p>The Poisson distribution is defined by a single parameter: the expected mean <span class="math inline">\(\lambda\)</span> (the Greek letter “lambda”). Noteworth, the variance of a Poisson distribution equals its mean.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" data-line-number="1"><span class="co">## Select a kmer and its counts</span></a>
<a class="sourceLine" id="cb63-2" data-line-number="2">kmer &lt;-<span class="st"> &quot;GA&quot;</span> <span class="co">## Select an arbitrary k-mer</span></a>
<a class="sourceLine" id="cb63-3" data-line-number="3">counts &lt;-<span class="st"> </span>kmer.table[, kmer]</a>
<a class="sourceLine" id="cb63-4" data-line-number="4">ci &lt;-<span class="st"> </span><span class="dv">2</span></a>
<a class="sourceLine" id="cb63-5" data-line-number="5"></a>
<a class="sourceLine" id="cb63-6" data-line-number="6"><span class="co">## Estimate lambda (mean of the Poisson) as the mean of the counts</span></a>
<a class="sourceLine" id="cb63-7" data-line-number="7">lambda &lt;-<span class="st"> </span><span class="kw">mean</span>(counts)</a>
<a class="sourceLine" id="cb63-8" data-line-number="8"></a>
<a class="sourceLine" id="cb63-9" data-line-number="9"><span class="co">## Compute the Poisson Probability Mass Function</span></a>
<a class="sourceLine" id="cb63-10" data-line-number="10">X &lt;-<span class="st"> </span><span class="dv">0</span><span class="op">:</span><span class="kw">max</span>(counts) <span class="co">## Compute the values covered by the k-mer counts</span></a>
<a class="sourceLine" id="cb63-11" data-line-number="11"><span class="co"># length(X)</span></a>
<a class="sourceLine" id="cb63-12" data-line-number="12">poisson.pmf &lt;-<span class="st"> </span><span class="kw">dpois</span>(<span class="dt">x =</span> X, <span class="dt">lambda =</span> lambda)</a>
<a class="sourceLine" id="cb63-13" data-line-number="13"><span class="co"># sum(poisson.pmf) # Validation : this should approximte 1</span></a>
<a class="sourceLine" id="cb63-14" data-line-number="14"></a>
<a class="sourceLine" id="cb63-15" data-line-number="15"><span class="co">## Compute expected number of genes for each count value</span></a>
<a class="sourceLine" id="cb63-16" data-line-number="16">poisson.exp &lt;-<span class="st"> </span>poisson.pmf <span class="op">*</span><span class="st"> </span>nb.genes</a>
<a class="sourceLine" id="cb63-17" data-line-number="17"><span class="co"># sum(poisson.exp) # Validation: this should approximate n</span></a></code></pre></div>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb64-1" data-line-number="1"><span class="co">## plot the histogram</span></a>
<a class="sourceLine" id="cb64-2" data-line-number="2">h &lt;-<span class="st"> </span><span class="kw">kmer.hist</span>(counts, kmer, </a>
<a class="sourceLine" id="cb64-3" data-line-number="3">               <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="kw">max</span>(poisson.exp)), </a>
<a class="sourceLine" id="cb64-4" data-line-number="4">               <span class="dt">class.interval =</span> <span class="dv">1</span>, <span class="dt">col =</span> <span class="st">&quot;#BBBBBB&quot;</span>, <span class="dt">border =</span> <span class="st">&quot;#BBBBBB&quot;</span>,</a>
<a class="sourceLine" id="cb64-5" data-line-number="5">               <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Poisson fit on &quot;</span>, kmer, <span class="st">&quot; counts&quot;</span>))</a>
<a class="sourceLine" id="cb64-6" data-line-number="6"><span class="kw">lines</span>(X, poisson.exp, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_fitting_plot_fp-1.png" alt="Fitting of a Poisson distribution on dinucleotide counts" width="60%" />
<p class="caption">
Fitting of a Poisson distribution on dinucleotide counts
</p>
</div>
<p>An alternative is to draw the lines as a histogram rather than a polygon frequency.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" data-line-number="1"><span class="co">## plot the histogram</span></a>
<a class="sourceLine" id="cb65-2" data-line-number="2">h &lt;-<span class="st"> </span><span class="kw">kmer.hist</span>(counts, kmer, </a>
<a class="sourceLine" id="cb65-3" data-line-number="3">               <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="kw">max</span>(poisson.exp)), </a>
<a class="sourceLine" id="cb65-4" data-line-number="4">               <span class="dt">class.interval =</span> <span class="dv">1</span>, <span class="dt">col =</span> <span class="st">&quot;#BBBBBB&quot;</span>, <span class="dt">border =</span> <span class="st">&quot;#BBBBBB&quot;</span>,</a>
<a class="sourceLine" id="cb65-5" data-line-number="5">               <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Poisson fit on &quot;</span>, kmer, <span class="st">&quot; counts&quot;</span>))</a>
<a class="sourceLine" id="cb65-6" data-line-number="6"><span class="kw">lines</span>(X, poisson.exp, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="dv">1</span>, <span class="dt">type =</span> <span class="st">&quot;h&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_fitting_plot_h-1.png" alt="Fitting of a Poisson distribution on dinucleotide counts" width="60%" />
<p class="caption">
Fitting of a Poisson distribution on dinucleotide counts
</p>
</div>
<p>We can now generate the same drawing for each k-mer.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb66-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">4</span>))</a>
<a class="sourceLine" id="cb66-2" data-line-number="2"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {</a>
<a class="sourceLine" id="cb66-3" data-line-number="3">  kmer &lt;-<span class="st"> </span><span class="kw">names</span>(kmer.table)[i]</a>
<a class="sourceLine" id="cb66-4" data-line-number="4">  counts &lt;-<span class="st"> </span>kmer.table[, kmer]</a>
<a class="sourceLine" id="cb66-5" data-line-number="5">  <span class="co">## plot the histogram</span></a>
<a class="sourceLine" id="cb66-6" data-line-number="6">  h &lt;-<span class="st"> </span><span class="kw">kmer.hist</span>(counts, kmer, </a>
<a class="sourceLine" id="cb66-7" data-line-number="7">                 <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="kw">max</span>(poisson.exp)), </a>
<a class="sourceLine" id="cb66-8" data-line-number="8">                 <span class="dt">class.interval =</span> <span class="dv">1</span>, <span class="dt">col =</span> <span class="st">&quot;#BBBBBB&quot;</span>, <span class="dt">border =</span> <span class="st">&quot;#BBBBBB&quot;</span>,</a>
<a class="sourceLine" id="cb66-9" data-line-number="9">                 <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Poisson fit on &quot;</span>, kmer, <span class="st">&quot; counts&quot;</span>))</a>
<a class="sourceLine" id="cb66-10" data-line-number="10">  <span class="kw">lines</span>(X, poisson.exp, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="fl">0.5</span>, <span class="dt">type =</span> <span class="st">&quot;h&quot;</span>)</a>
<a class="sourceLine" id="cb66-11" data-line-number="11"></a>
<a class="sourceLine" id="cb66-12" data-line-number="12">}</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/02_combinatorix_poisson_fit_all_kmers-1.png" alt="Poisson fit on dinucleotide counts from Human promoter sequences. " width="100%" />
<p class="caption">
Poisson fit on dinucleotide counts from Human promoter sequences.
</p>
</div>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</a></code></pre></div>
</div>
<div id="q-q-plots" class="section level3">
<h3>Q-Q plots</h3>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb68-1" data-line-number="1">kmer &lt;-<span class="st"> &quot;TT&quot;</span></a>
<a class="sourceLine" id="cb68-2" data-line-number="2">counts &lt;-<span class="st"> </span>kmer.table[, kmer]</a>
<a class="sourceLine" id="cb68-3" data-line-number="3">lambda &lt;-<span class="st"> </span><span class="kw">mean</span>(counts)</a>
<a class="sourceLine" id="cb68-4" data-line-number="4">X &lt;-<span class="st"> </span><span class="dv">0</span><span class="op">:</span>(<span class="kw">max</span>(counts))</a>
<a class="sourceLine" id="cb68-5" data-line-number="5"></a>
<a class="sourceLine" id="cb68-6" data-line-number="6"><span class="co">## Generate random counts that follow a Poisson loaw</span></a>
<a class="sourceLine" id="cb68-7" data-line-number="7">rcounts &lt;-<span class="st">  </span><span class="kw">rpois</span>(<span class="dt">n =</span> <span class="kw">length</span>(counts), <span class="dt">lambda =</span> lambda)</a>
<a class="sourceLine" id="cb68-8" data-line-number="8"></a>
<a class="sourceLine" id="cb68-9" data-line-number="9"><span class="co">## Draw a Q-Q plot with the random counts versus actual counts</span></a>
<a class="sourceLine" id="cb68-10" data-line-number="10"><span class="kw">qqplot</span>(<span class="dt">x =</span> counts, <span class="dt">y =</span> rcounts, </a>
<a class="sourceLine" id="cb68-11" data-line-number="11">       <span class="dt">main =</span> <span class="st">&quot;QQplot - counts vs random Poisson&quot;</span>,</a>
<a class="sourceLine" id="cb68-12" data-line-number="12">       <span class="dt">col =</span> <span class="st">&quot;#88AABB&quot;</span>,</a>
<a class="sourceLine" id="cb68-13" data-line-number="13">       <span class="dt">xlab =</span> <span class="st">&quot;Counts&quot;</span>,</a>
<a class="sourceLine" id="cb68-14" data-line-number="14">       <span class="dt">ylab =</span> <span class="st">&quot;Random counts (rpois)&quot;</span>)</a>
<a class="sourceLine" id="cb68-15" data-line-number="15"><span class="kw">grid</span>()</a>
<a class="sourceLine" id="cb68-16" data-line-number="16"><span class="kw">abline</span>(<span class="dt">a =</span> <span class="dv">0</span>, <span class="dt">b =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><img src="figures/02_combinatorix_unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="fitting-a-binomial-distribution" class="section level3">
<h3>Fitting a binomial distribution</h3>
</div>
<div id="fitting-a-normal-distribution" class="section level3">
<h3>Fitting a normal distribution</h3>
</div>
<div id="fitting-a-negative-binomial" class="section level3">
<h3>Fitting a negative binomial</h3>
<ol start="3" style="list-style-type: decimal">
<li>Estimate the goodness of fit for these distributions.</li>
</ol>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>Note: normally it is recommended to check this option, but we intently inactivate it in order to get sequences of the same sizes. <a href="#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
