<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2019-09-20" />

<title>Analysis of a genome annotation table</title>

<script src="02_yeast_annotations_EN_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="02_yeast_annotations_EN_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="02_yeast_annotations_EN_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="02_yeast_annotations_EN_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="02_yeast_annotations_EN_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="02_yeast_annotations_EN_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="02_yeast_annotations_EN_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="02_yeast_annotations_EN_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="02_yeast_annotations_EN_files/navigation-1.1/tabsets.js"></script>
<script src="02_yeast_annotations_EN_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #303030; color: #cccccc; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #cccccc; background-color: #303030; }
code > span.kw { color: #f0dfaf; } /* Keyword */
code > span.dt { color: #dfdfbf; } /* DataType */
code > span.dv { color: #dcdccc; } /* DecVal */
code > span.bn { color: #dca3a3; } /* BaseN */
code > span.fl { color: #c0bed1; } /* Float */
code > span.ch { color: #dca3a3; } /* Char */
code > span.st { color: #cc9393; } /* String */
code > span.co { color: #7f9f7f; } /* Comment */
code > span.ot { color: #efef8f; } /* Other */
code > span.al { color: #ffcfaf; } /* Alert */
code > span.fu { color: #efef8f; } /* Function */
code > span.er { color: #c3bf9f; } /* Error */
code > span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */
code > span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code > span.sc { color: #dca3a3; } /* SpecialChar */
code > span.vs { color: #cc9393; } /* VerbatimString */
code > span.ss { color: #cc9393; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #f0dfaf; } /* ControlFlow */
code > span.op { color: #f0efd0; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code > span.at { } /* Attribute */
code > span.do { color: #7f9f7f; } /* Documentation */
code > span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code > span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code > span.in { color: #7f9f7f; font-weight: bold; } /* Information */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Analysis of a genome annotation table</h1>
<h3 class="subtitle">Probabilities and statistics for biology (STAT1)</h3>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">2019-09-20</h4>

</div>


<div id="goal-of-this-practical" class="section level2">
<h2>Goal of this practical</h2>
<p>During this practical session, you will run the following tasks:</p>
<ol style="list-style-type: decimal">
<li>Handle a table containing annotated features of the yeast genome.</li>
<li>Select a subset of the data by filtering rows based on a given criterion (annotation type, chromosome, …)</li>
<li>Generate graphics to represent different aspects of the data.</li>
<li>Compute estimators of central tendency and dispersion.</li>
<li>Compute a confidence interval around the mean.</li>
</ol>
</div>
<div id="expected-report" class="section level2">
<h2>Expected report</h2>
<p>At the end of the practical you will be asked to submit two documents</p>
<ol style="list-style-type: decimal">
<li>Your <strong>R code</strong>.Each question must be explicitly formulated before presenting the results that answer it and giving an interpretation of these results.</li>
<li>UA <strong>synthetic report</strong>, which will include a presentation of the main results (figures, descriptive stats, tables) as well as your interpretation of the result.</li>
</ol>
<div id="expectation-for-the-code" class="section level3">
<h3>Expectation for the code</h3>
<ol style="list-style-type: decimal">
<li><p>The code must be <strong>readable and undestandable</strong>: choose variable names that explicitly indicate what they represent.</p></li>
<li><p>The code must be properly documented (the <code>#</code> symbol starts a comment, either at the beginning or in the middle of a line of code).</p>
<ul>
<li><p>Before each chunk of code, explain what this code is supposed to do, what it serves to.</p></li>
<li><p>Don’t hesitate to occasionally add some comment words to justify the chosen approach.</p></li>
<li><p>Each time you define a variable, add a comment on the same line to indicate what this variable represents.</p></li>
</ul></li>
<li><p>The code must be <strong>portable</strong>: other people should be able to download it and run it on their computer. For this practical, I will systematically test whether your code can run on my computer. hard-coded absolute paths of a file on your machine should thus always be avoided (we will indicate hereafter how to define relative paths relative to the root of your user account).</p></li>
</ol>
</div>
<div id="expected-content-for-the-interpretation-report" class="section level3">
<h3>Expected content for the interpretation report</h3>
<p>Your report must be synthetic (1 text page max + as many figures and table as you wish)</p>
<p>Each question must be explicitly formulated before presenting the results that answer it and then interpreting those results.</p>
<p>Each figure or table must be documented with a legend that allows a naive reader to understand what it represents. The interpretation of the results displayed on a figure or table will be found in the main text (with a reference to the figure or table number).</p>
</div>
</div>
<div id="historical-example-yeast-genome" class="section level2">
<h2>Historical example: yeast genome</h2>
<ul>
<li>1992: publication of the first complete eukaryotic chromosome, the 3rd yeast chromosome.</li>
<li>1996: publication of the complete genome.</li>
</ul>
<p>On the base of the genes of the 3rd chromosome (sample) we can estimate the average size of a yeast gene.</p>
<p><strong>Questions: </strong></p>
<ol style="list-style-type: lower-alpha">
<li><p>Would the sample mean (chromosome III) be sufficient to predict the population mean (complete genome) ?</p>
<p>To answer this question, we will imagine that we came back in 1992, and will use all the genes of chromosome III (considered here as a sample of the genome) to estimate the average size of genes for the whole genome (the “population” of genes“).</p></li>
<li><p>Can this sample be described as “simple and independent” ?</p></li>
</ol>
</div>
<div id="analysis-of-the-length-of-the-bakers-yeast-genes" class="section level2">
<h2>Analysis of the length of the baker’s yeast genes</h2>
<div class="figure" style="text-align: center">
<img src="figures/cds_length_histo-1.png" alt="Distribution of cds lengths for Saccharomyces cerevisiae. " width="90%" />
<p class="caption">
Distribution of cds lengths for Saccharomyces cerevisiae.
</p>
</div>
</div>
<div id="tutorial" class="section level2">
<h2>Tutorial</h2>
<p>Before moving to the exercises, we show you here some basic elements about reading, manipulating and writing data tables with R.</p>
<div id="the-path-to-the-home-manual" class="section level3">
<h3>The path to the home (manual)</h3>
<p>We will create a folder for this tutorial, starting from the root of our account.</p>
<p>First possibility (<strong>quick but not very elegant</strong>): enter (manually) the path from the root of your account in a variable</p>
<p><code>dir.home &lt;- /the/path/to/the/home</code></p>
<ul>
<li>Advantage: fast and convenient</li>
<li>Disadvantage: not portable, will only work on your computer</li>
</ul>
</div>
<div id="the-path-to-the-home-automatic" class="section level3">
<h3>The path to the home (automatic)</h3>
<p>A more general solution: use the <strong>R</strong> command <code>Sys.getenv()</code>.</p>
<ul>
<li>Invoked without parameters, this command lists all environment variables (your system configuration).</li>
<li>The output can be restricted to a given environment variable, for example <code>Sys.getenv(&quot;HOME&quot;)</code> returns the path to the root of your account.</li>
</ul>
<p><strong>Note:</strong> equivalent writing with Linux: the tilde symbol <code>~</code> also indicates the path to the root of your account.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Identify the home directory 
## by getting the environment variable HOME 
dir.home &lt;-<span class="st"> </span><span class="kw">Sys.getenv</span>(<span class="st">&quot;HOME&quot;</span>)
<span class="kw">print</span>(dir.home)</code></pre></div>
<pre><code>[1] &quot;/home/khamvongsa&quot;</code></pre>
</div>
<div id="creating-a-folder-for-the-tp" class="section level3">
<h3>Creating a folder for the TP</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Define a variable containing the path of the results for this tutorial
dir.tuto &lt;-<span class="st"> </span><span class="kw">file.path</span>(dir.home, <span class="st">&quot;stat1&quot;</span>, <span class="st">&quot;TP2&quot;</span>)

<span class="kw">print</span>(dir.tuto)</code></pre></div>
<pre><code>[1] &quot;/home/khamvongsa/stat1/TP2&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Create the directory for this tutorial
<span class="kw">dir.create</span>(<span class="dt">path =</span> dir.tuto, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)

## Go to the tutorial directory
<span class="kw">setwd</span>(dir.tuto)

## List the files already present in the folder (if any)
<span class="kw">list.files</span>()</code></pre></div>
<pre><code>[1] &quot;3nt_genomic_Saccharomyces_cerevisiae-ovlp-1str.tab&quot;
[2] &quot;chrom_sizes.tsv&quot;                                   
[3] &quot;Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz&quot;        </code></pre>
</div>
<div id="downloading-the-gtf-file-from-ensemblgenomes" class="section level3">
<h3>Downloading the GTF file from EnsemblGenomes</h3>
<p><strong>Tips: </strong> before downloading the annotation file (GTF) from EnsemblGenomes to our computer, we will check if it is already present (and in this case we do not re-download it).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Define the URL of the annotation file (GTF-formatted)
gtf.URL &lt;-<span class="st"> &quot;ftp://ftp.ensemblgenomes.org/pub/release-37/fungi/gtf/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz&quot;</span>

## Define the path where the local copy will be stored
local.GTF &lt;-<span class="st"> </span><span class="kw">file.path</span>(dir.tuto, <span class="st">&quot;Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz&quot;</span>)

## If the local file file laready exists, skip the download
<span class="cf">if</span> (<span class="kw">file.exists</span>(local.GTF)) {
  <span class="kw">message</span>(<span class="st">&quot;GTF file already exists in the tutorial folder: &quot;</span>, local.GTF)
} <span class="cf">else</span> {
  ## Download annotation table in GTF format
  <span class="kw">download.file</span>(<span class="dt">url =</span> gtf.URL, <span class="dt">destfile =</span> local.GTF)
}</code></pre></div>
</div>
<div id="loading-a-data-table" class="section level3">
<h3>Loading a data table</h3>
<p>R has several types of tabular structures (matrix, data.frame, table).</p>
<p>The most commonly used structure is the <code>data.frame</code>, which consists of an array of values (numeric or strings) whose rows and columns are associated with names.</p>
<p>The function <code>read.table()</code> allows you to read a text file containing a data table, and store the content in a variable.</p>
<p>Several functions derived from <code>read.table()</code> make it easier to read different types of formats:</p>
<ul>
<li><code>read.delim()</code> for files whose columns are delimited by a particular character (usually the tab, represented by “”).</li>
<li><code>read.csv()</code> for files “comma-separated values”.</li>
</ul>
<ol style="list-style-type: decimal">
<li>Download the following file to your computer:</li>
</ol>
<ul>
<li><a href="../../data/Saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.37.gtf">Saccharomyces_cerevisiae.R64-1-1.37.gtf</a></li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>Load it using the read.table function (for this you must replace the path below by that of your computer).</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Read a GTF file with yeast genome annotations

## Load the feature table
feature.table &lt;-<span class="st"> </span><span class="kw">read.table</span>(
  local.GTF, 
  <span class="dt">comment.char =</span> <span class="st">&quot;#&quot;</span>, 
  <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, 
  <span class="dt">header=</span><span class="ot">FALSE</span>, 
  <span class="dt">row.names=</span><span class="ot">NULL</span>)

## The bed format does not contain any column header, 
## so we set it manually based on the description of the format, 
## found here: 
##     http://www.ensembl.org/info/website/upload/gff.html
<span class="kw">names</span>(feature.table) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;seqname&quot;</span>, <span class="st">&quot;source&quot;</span>, <span class="st">&quot;feature&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;end&quot;</span>, <span class="st">&quot;score&quot;</span>, <span class="st">&quot;strand&quot;</span>, <span class="st">&quot;frame&quot;</span>, <span class="st">&quot;attribute&quot;</span>)</code></pre></div>
</div>
<div id="exploring-the-content-of-a-data-table" class="section level3">
<h3>Exploring the content of a data table</h3>
<p>The first thing to do after loading a data table is to check its dimensions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(feature.table) ## Dimensions of the tbale</code></pre></div>
<pre><code>[1] 43028     9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(feature.table) ## Number of rows</code></pre></div>
<pre><code>[1] 43028</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ncol</span>(feature.table) ## Number of columns</code></pre></div>
<pre><code>[1] 9</code></pre>
<p>The display of the complete annotation table would not be very readable, since it contains tens of thousands of lines.</p>
<p>We can display the first lines with the function <code>head()</code>.</p>
<p><strong>Note: </strong> the last column is particularly heavy (it contains a lot of information). We will see later how to select a subset of the columns to simplify the display.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Display the 5 first rows of the feature table
<span class="kw">head</span>(feature.table, <span class="dt">n =</span> <span class="dv">5</span>) </code></pre></div>
<pre><code>  seqname source     feature start  end score strand frame
1      IV    SGD        gene  1802 2953     .      +     .
2      IV    SGD  transcript  1802 2953     .      +     .
3      IV    SGD        exon  1802 2953     .      +     .
4      IV    SGD         CDS  1802 2950     .      +     0
5      IV    SGD start_codon  1802 1804     .      +     0
                                                                                                                                                                                                                                    attribute
1                                                                                                                                                              gene_id YDL248W; gene_name COS7; gene_source SGD; gene_biotype protein_coding;
2                                                       gene_id YDL248W; transcript_id YDL248W; gene_name COS7; gene_source SGD; gene_biotype protein_coding; transcript_name COS7; transcript_source SGD; transcript_biotype protein_coding;
3                     gene_id YDL248W; transcript_id YDL248W; exon_number 1; gene_name COS7; gene_source SGD; gene_biotype protein_coding; transcript_name COS7; transcript_source SGD; transcript_biotype protein_coding; exon_id YDL248W.1;
4 gene_id YDL248W; transcript_id YDL248W; exon_number 1; gene_name COS7; gene_source SGD; gene_biotype protein_coding; transcript_name COS7; transcript_source SGD; transcript_biotype protein_coding; protein_id YDL248W; protein_version 1;
5                                        gene_id YDL248W; transcript_id YDL248W; exon_number 1; gene_name COS7; gene_source SGD; gene_biotype protein_coding; transcript_name COS7; transcript_source SGD; transcript_biotype protein_coding;</code></pre>
<p>The function <code>tail()</code> displays the last few lines:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Display the 5 last rows of the feature table
<span class="kw">tail</span>(feature.table, <span class="dt">n =</span> <span class="dv">5</span>) </code></pre></div>
<pre><code>      seqname source     feature start   end score strand frame
43024    Mito    SGD  transcript 85554 85709     .      +     .
43025    Mito    SGD        exon 85554 85709     .      +     .
43026    Mito    SGD         CDS 85554 85706     .      +     0
43027    Mito    SGD start_codon 85554 85556     .      +     0
43028    Mito    SGD  stop_codon 85707 85709     .      +     0
                                                                                                                                                                                            attribute
43024                                                     gene_id Q0297; transcript_id Q0297; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding;
43025                     gene_id Q0297; transcript_id Q0297; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding; exon_id Q0297.1;
43026 gene_id Q0297; transcript_id Q0297; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding; protein_id Q0297; protein_version 1;
43027                                      gene_id Q0297; transcript_id Q0297; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding;
43028                                      gene_id Q0297; transcript_id Q0297; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding;</code></pre>
<p>If you are using the <strong>RStudio</strong> environment, you can display the table in a dynamic viewer pane with the function <code>View()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## In RStudio, display the table in a separate tab
<span class="kw">View</span>(feature.table) </code></pre></div>
</div>
<div id="selection-of-subsets-from-a-table" class="section level3">
<h3>Selection of subsets from a table</h3>
<p>Selection of a line specified by its index.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">feature.table[<span class="dv">12</span>,]</code></pre></div>
<pre><code>   seqname source    feature start  end score strand frame
12      IV    SGD stop_codon  3834 3836     .      +     0
                                                                                                                                                            attribute
12 gene_id YDL247W-A; transcript_id YDL247W-A; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding;</code></pre>
<p>Selection of a column specified by its index (display of the first values only).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(feature.table[,<span class="dv">3</span>])</code></pre></div>
<pre><code>[1] gene        transcript  exon        CDS         start_codon stop_codon 
Levels: CDS exon gene start_codon stop_codon transcript</code></pre>
<p>Selection of a cell by combining row and column indices.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">feature.table[<span class="dv">12</span>, <span class="dv">3</span>]</code></pre></div>
<pre><code>[1] stop_codon
Levels: CDS exon gene start_codon stop_codon transcript</code></pre>
<p>Selection of a column and/or row set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">feature.table[<span class="dv">100</span><span class="op">:</span><span class="dv">105</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</code></pre></div>
<pre><code>    seqname source     feature start   end score
100      IV    SGD         CDS 34240 36477     .
101      IV    SGD start_codon 36475 36477     .
102      IV    SGD  stop_codon 34237 34239     .
103      IV    SGD        gene 36797 38173     .
104      IV    SGD  transcript 36797 38173     .
105      IV    SGD        exon 36797 38173     .</code></pre>
<p>Selection of specific columns (here, the genomic coordinates of each feature): chromosome, beginning, end, strand.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">feature.table[<span class="dv">100</span><span class="op">:</span><span class="dv">105</span>, <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">7</span>)]</code></pre></div>
<pre><code>    seqname start   end strand
100      IV 34240 36477      -
101      IV 36475 36477      -
102      IV 34237 34239      -
103      IV 36797 38173      +
104      IV 36797 38173      +
105      IV 36797 38173      +</code></pre>
<p>Select a column based on its name.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Select the &quot;start&quot; column and print the 100 first results
<span class="kw">head</span>(feature.table<span class="op">$</span>start, <span class="dt">n=</span><span class="dv">100</span>)</code></pre></div>
<pre><code>  [1]  1802  1802  1802  1802  1802  2951  3762  3762  3762  3762  3762
 [12]  3834  5985  5985  5985  5985  5985  7812  8683  8683  8683  8686
 [23]  9754  8683 11657 11657 11657 11660 13358 11657 16204 16204 16204
 [34] 16204 16204 17224 17577 17577 17577 17580 18564 17577 18959 18959
 [45] 18959 18959 18959 19310 20635 20635 20635 20635 20635 21004 22471
 [56] 22471 22471 22474 22606 22471 22823 22823 22823 22823 22823 25874
 [67] 26403 26403 26403 26406 28773 26403 28985 28985 28985 28988 30452
 [78] 28985 30657 30657 30657 30657 30657 31827 32296 32296 32296 32296
 [89] 32296 33232 33415 33415 33415 33418 33916 33415 34237 34237 34237
[100] 34240</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Print the 20 first values of the &quot;feature&quot; field, which indicates the feature type
<span class="kw">head</span>(feature.table<span class="op">$</span>feature, <span class="dt">n=</span><span class="dv">20</span>)</code></pre></div>
<pre><code> [1] gene        transcript  exon        CDS         start_codon
 [6] stop_codon  gene        transcript  exon        CDS        
[11] start_codon stop_codon  gene        transcript  exon       
[16] CDS         start_codon stop_codon  gene        transcript 
Levels: CDS exon gene start_codon stop_codon transcript</code></pre>
<p>Selection of several columns based on their names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Select the &quot;start&quot; column and print the 100 first results
feature.table[<span class="dv">100</span><span class="op">:</span><span class="dv">106</span>, <span class="kw">c</span>(<span class="st">&quot;seqname&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;end&quot;</span>, <span class="st">&quot;strand&quot;</span>)]</code></pre></div>
<pre><code>    seqname start   end strand
100      IV 34240 36477      -
101      IV 36475 36477      -
102      IV 34237 34239      -
103      IV 36797 38173      +
104      IV 36797 38173      +
105      IV 36797 38173      +
106      IV 36797 38170      +</code></pre>
<p><strong>Note</strong>: Selection of several columns based on their names. It is also possible to name the rows of a data.frame but the GTF table does not support this. We will see more examples later.</p>
</div>
<div id="selection-of-a-subset-of-rows-based-on-the-content-of-a-column" class="section level3">
<h3>Selection of a subset of rows based on the content of a column</h3>
<p>The function <code>subset()</code> allows you to select a subset of the rows of a data.frame based on a condition applied to one or more columns.</p>
<p>We can apply it to select the subset of rows in the annotation table corresponding to coding sequences (CDS).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Select subset of features having &quot;cds&quot; as &quot;feature&quot; attribute
cds &lt;-<span class="st"> </span><span class="kw">subset</span>(feature.table, feature<span class="op">==</span><span class="st">&quot;cds&quot;</span>)

<span class="kw">nrow</span>(feature.table) ## Count the number of features</code></pre></div>
<pre><code>[1] 43028</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(cds) ## Count the number of cds</code></pre></div>
<pre><code>[1] 0</code></pre>
</div>
<div id="count-by-value" class="section level3">
<h3>Count by value</h3>
<p>The function <code>table()</code> allows you to count the occurrences of each value in a vector or array. Some examples of use below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Count the number of featues per chromosome
<span class="kw">table</span>(feature.table<span class="op">$</span>seqname)</code></pre></div>
<pre><code>
   I   II  III   IV   IX Mito    V   VI  VII VIII    X   XI  XII XIII  XIV 
 759 2912 1210 5374 1567  327 2159  946 3856 2054 2617 2231 3789 3311 2774 
  XV  XVI 
3846 3296 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Count the number of features per type
<span class="kw">table</span>(feature.table<span class="op">$</span>feature)</code></pre></div>
<pre><code>
        CDS        exon        gene start_codon  stop_codon  transcript 
       7050        7872        7445        6700        6516        7445 </code></pre>
<p>Contingency tables can be calculated by counting the number of combinations between 2 vectors (or 2 columns of a table).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">##  Table with two vectors
<span class="kw">table</span>(feature.table<span class="op">$</span>feature, feature.table<span class="op">$</span>seqname)</code></pre></div>
<pre><code>             
                I  II III  IV  IX Mito   V  VI VII VIII   X  XI XII XIII
  CDS         122 492 194 895 255   59 345 151 619  346 422 361 615  544
  exon        137 525 224 961 288   94 400 180 710  373 480 404 698  610
  gene        132 494 213 914 274   62 383 167 676  349 458 388 658  573
  start_codon 119 464 185 853 243   28 328 143 593  325 406 348 586  514
  stop_codon  117 443 181 837 233   22 320 138 582  312 393 342 574  497
  transcript  132 494 213 914 274   62 383 167 676  349 458 388 658  573
             
              XIV  XV XVI
  CDS         458 623 549
  exon        500 689 599
  gene        475 665 564
  start_codon 438 607 520
  stop_codon  428 597 500
  transcript  475 665 564</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Same result with a 2-column data frame
<span class="kw">table</span>(feature.table[, <span class="kw">c</span>(<span class="st">&quot;feature&quot;</span>, <span class="st">&quot;seqname&quot;</span>)])</code></pre></div>
<pre><code>             seqname
feature         I  II III  IV  IX Mito   V  VI VII VIII   X  XI XII XIII
  CDS         122 492 194 895 255   59 345 151 619  346 422 361 615  544
  exon        137 525 224 961 288   94 400 180 710  373 480 404 698  610
  gene        132 494 213 914 274   62 383 167 676  349 458 388 658  573
  start_codon 119 464 185 853 243   28 328 143 593  325 406 348 586  514
  stop_codon  117 443 181 837 233   22 320 138 582  312 393 342 574  497
  transcript  132 494 213 914 274   62 383 167 676  349 458 388 658  573
             seqname
feature       XIV  XV XVI
  CDS         458 623 549
  exon        500 689 599
  gene        475 665 564
  start_codon 438 607 520
  stop_codon  428 597 500
  transcript  475 665 564</code></pre>
</div>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<div id="gtf-format-specifications" class="section level3">
<h3>1. GTF format specifications</h3>
<p>Read the GTF format specifications.</p>
<ul>
<li>Ensembl (<a href="http://www.ensembl.org/info/website/upload/gff.html" class="uri">http://www.ensembl.org/info/website/upload/gff.html</a>)</li>
<li>UCSC (<a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format4" class="uri">https://genome.ucsc.edu/FAQ/FAQformat.html#format4</a>)</li>
</ul>
</div>
<div id="creating-a-local-folder-for-the-tp" class="section level3">
<h3>2. Creating a local folder for the TP</h3>
<p>Create a local folder (for example: <code>stat1/TP_yeast</code> from the root of your account). We suggest you to use the following functions:</p>
<ul>
<li><code>Sys.getenv(&quot;HOME&quot;)</code> (Linux and Mac OS X), to get the root of your user account;</li>
<li><code>file.path()</code> to build a path;</li>
<li><code>dir.create()</code> to create the folder for the TP. Read carefully the options of this function with <code>help(dir.create)</code></li>
</ul>
</div>
<div id="locating-the-annotation-file" class="section level3">
<h3>3. Locating the annotation file</h3>
<p>Locate the yeast genome annotation file in GTF format in this local folder.</p>
<ul>
<li>Site Ensembl Fungi: <a href="http://fungi.ensembl.org/" class="uri">http://fungi.ensembl.org/</a></li>
<li>Click “Downloads” to access the <a href="http://fungi.ensembl.org/info/website/ftp/">ftp website</a></li>
<li>In the search box, type <em>“saccharomyces cerevisiae”</em> and follow the link “<a href="ftp://ftp.ensemblgenomes.org/pub/release-37/fungi/gtf/saccharomyces_cerevisiae">GTF</a>”</li>
<li>Copy the address (URL) of the file <a href="ftp://ftp.ensemblgenomes.org/pub/release-37/fungi/gtf/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz">Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz</a></li>
</ul>
</div>
<div id="downloading-a-file-from-an-ftp-website" class="section level3">
<h3>4. Downloading a file from an ftp website</h3>
<p>Suggested functions:</p>
<ul>
<li><code>download.file()</code> (read the help to know the arguments)</li>
</ul>
</div>
<div id="loading-a-data-table-in-r" class="section level3">
<h3>5. Loading a data table in R</h3>
<p>Write a script that loads the data table into a variable named <code>feature.table</code>, using the function R <code>read.delim()</code>.</p>
<p>Be sure to ignore the comment lines (which start with a character <code>#</code>).</p>
</div>
<div id="compute-the-length-of-coding-genes" class="section level3">
<h3>6. Compute the length of coding genes</h3>
<ul>
<li>Add to the annotation table (<code>feature.table</code>) a column entitled “length” which indicates the length of each annotated genomic feature.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Add a colmn with feature lengths
feature.table[, <span class="st">&quot;length&quot;</span>] &lt;-<span class="st"> </span>feature.table[, <span class="st">&quot;end&quot;</span>] <span class="op">-</span><span class="st"> </span>feature.table[, <span class="st">&quot;start&quot;</span>] <span class="op">+</span><span class="st"> </span><span class="dv">1</span>

## Add a colmn with feature lengths: equivalent result with simpler notation
feature.table<span class="op">$</span>length &lt;-<span class="st"> </span>feature.table<span class="op">$</span>end <span class="op">-</span><span class="st"> </span>feature.table<span class="op">$</span>start <span class="op">+</span><span class="st"> </span><span class="dv">1</span></code></pre></div>
<ul>
<li><p>Count the number of rows in the table corresponding to each type of annotation (3rd column of the GTF, “feature”).</p>
<ul>
<li>fonction <code>table()</code></li>
</ul></li>
</ul>
<pre><code>~table(feature.table$feature)</code></pre>
<ul>
<li><p>Select the lines corresponding to coding regions (“CDS”)</p>
<ul>
<li>fonction <code>subset()</code></li>
</ul></li>
<li><p>Count the number of CDS per chromosome.</p>
<ul>
<li>fonction <code>table()</code></li>
</ul></li>
</ul>
<pre><code>
   I   II  III   IV   IX Mito    V   VI  VII VIII    X   XI  XII XIII  XIV 
 122  492  194  895  255   59  345  151  619  346  422  361  615  544  458 
  XV  XVI 
 623  549 </code></pre>
<ul>
<li>Load the chromosome size table <a href="../../data/Saccharomyces_cerevisiae/chrom_sizes.tsv">chrom_sizes.tsv</a>, and calculate the gene density for each chromosome (number of genes per Mb).</li>
</ul>
<pre><code>[1] 316617</code></pre>
</div>
<div id="histogram-of-gene-length" class="section level3">
<h3>6. Histogram of gene length</h3>
<p>By using the function <code>hist()</code>, draw a histogram representing the length distribution of the CDS.</p>
<p><img src="figures/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Choose the class intervals in a way that the histogram is informative (neither too large nor too few classes).</p>
<p><img src="figures/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Retrieve the result of <code>hist()</code> in a variable named <code>cds.length.hist</code>.</p>
<p><img src="figures/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Print the result on the screen (<code>print()</code>) and analyze the structure of the variable <code>cds.length.hist</code> (this is a list variable).</p>
<p>Useful functions:</p>
<pre><code>$breaks
  [1]     0   100   200   300   400   500   600   700   800   900  1000
 [12]  1100  1200  1300  1400  1500  1600  1700  1800  1900  2000  2100
 [23]  2200  2300  2400  2500  2600  2700  2800  2900  3000  3100  3200
 [34]  3300  3400  3500  3600  3700  3800  3900  4000  4100  4200  4300
 [45]  4400  4500  4600  4700  4800  4900  5000  5100  5200  5300  5400
 [56]  5500  5600  5700  5800  5900  6000  6100  6200  6300  6400  6500
 [67]  6600  6700  6800  6900  7000  7100  7200  7300  7400  7500  7600
 [78]  7700  7800  7900  8000  8100  8200  8300  8400  8500  8600  8700
 [89]  8800  8900  9000  9100  9200  9300  9400  9500  9600  9700  9800
[100]  9900 10000 10100 10200 10300 10400 10500 10600 10700 10800 10900
[111] 11000 11100 11200 11300 11400 11500 11600 11700 11800 11900 12000
[122] 12100 12200 12300 12400 12500 12600 12700 12800 12900 13000 13100
[133] 13200 13300 13400 13500 13600 13700 13800 13900 14000 14100 14200
[144] 14300 14400 14500 14600 14700 14800

$counts
  [1] 288 236 292 689 394 322 331 282 317 300 295 292 250 316 233 215 194
 [18] 190 170 128 131 109  99  81  81  72  80  51  53  39  39  38  45  40
 [35]  27  28  26  14  17  40  27  12  14  17  12  13  11  10   4  11   6
 [52]   3   3   7   1   4   8   3   2   1   3   2   0   2   3   3   4   0
 [69]   1   0   0   2   1   0   4   0   0   0   1   0   1   0   1   1   0
 [86]   0   0   0   1   0   0   0   2   0   1   0   0   0   1   0   0   0
[103]   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0
[120]   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0
[137]   0   0   0   0   0   0   0   0   0   0   0   1

$density
  [1] 4.085106e-04 3.347518e-04 4.141844e-04 9.773050e-04 5.588652e-04
  [6] 4.567376e-04 4.695035e-04 4.000000e-04 4.496454e-04 4.255319e-04
 [11] 4.184397e-04 4.141844e-04 3.546099e-04 4.482270e-04 3.304965e-04
 [16] 3.049645e-04 2.751773e-04 2.695035e-04 2.411348e-04 1.815603e-04
 [21] 1.858156e-04 1.546099e-04 1.404255e-04 1.148936e-04 1.148936e-04
 [26] 1.021277e-04 1.134752e-04 7.234043e-05 7.517730e-05 5.531915e-05
 [31] 5.531915e-05 5.390071e-05 6.382979e-05 5.673759e-05 3.829787e-05
 [36] 3.971631e-05 3.687943e-05 1.985816e-05 2.411348e-05 5.673759e-05
 [41] 3.829787e-05 1.702128e-05 1.985816e-05 2.411348e-05 1.702128e-05
 [46] 1.843972e-05 1.560284e-05 1.418440e-05 5.673759e-06 1.560284e-05
 [51] 8.510638e-06 4.255319e-06 4.255319e-06 9.929078e-06 1.418440e-06
 [56] 5.673759e-06 1.134752e-05 4.255319e-06 2.836879e-06 1.418440e-06
 [61] 4.255319e-06 2.836879e-06 0.000000e+00 2.836879e-06 4.255319e-06
 [66] 4.255319e-06 5.673759e-06 0.000000e+00 1.418440e-06 0.000000e+00
 [71] 0.000000e+00 2.836879e-06 1.418440e-06 0.000000e+00 5.673759e-06
 [76] 0.000000e+00 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00
 [81] 1.418440e-06 0.000000e+00 1.418440e-06 1.418440e-06 0.000000e+00
 [86] 0.000000e+00 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00
 [91] 0.000000e+00 0.000000e+00 2.836879e-06 0.000000e+00 1.418440e-06
 [96] 0.000000e+00 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00
[101] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[106] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[111] 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00 0.000000e+00
[116] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[121] 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00 0.000000e+00
[126] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[131] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[136] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[141] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[146] 0.000000e+00 0.000000e+00 1.418440e-06

$mids
  [1]    50   150   250   350   450   550   650   750   850   950  1050
 [12]  1150  1250  1350  1450  1550  1650  1750  1850  1950  2050  2150
 [23]  2250  2350  2450  2550  2650  2750  2850  2950  3050  3150  3250
 [34]  3350  3450  3550  3650  3750  3850  3950  4050  4150  4250  4350
 [45]  4450  4550  4650  4750  4850  4950  5050  5150  5250  5350  5450
 [56]  5550  5650  5750  5850  5950  6050  6150  6250  6350  6450  6550
 [67]  6650  6750  6850  6950  7050  7150  7250  7350  7450  7550  7650
 [78]  7750  7850  7950  8050  8150  8250  8350  8450  8550  8650  8750
 [89]  8850  8950  9050  9150  9250  9350  9450  9550  9650  9750  9850
[100]  9950 10050 10150 10250 10350 10450 10550 10650 10750 10850 10950
[111] 11050 11150 11250 11350 11450 11550 11650 11750 11850 11950 12050
[122] 12150 12250 12350 12450 12550 12650 12750 12850 12950 13050 13150
[133] 13250 13350 13450 13550 13650 13750 13850 13950 14050 14150 14250
[144] 14350 14450 14550 14650 14750

$xname
[1] &quot;cds$length&quot;

$equidist
[1] TRUE

attr(,&quot;class&quot;)
[1] &quot;histogram&quot;</code></pre>
<ul>
<li><code>class(cds.length.hist)</code></li>
<li><code>attributes(cds.length.hist)</code></li>
</ul>
<p>Other types of graphs allow you to explore the distribution of a set of data. In particular, box plots display, for a series of data, the median, the quarterfinal range, a confidence interval and outliers.</p>
<div class="figure" style="text-align: center">
<img src="figures/unnamed-chunk-24-1.png" alt="Boîte à moustache indiquant la distribution de longueur des gènes par chromosome. " width="672" />
<p class="caption">
Boîte à moustache indiquant la distribution de longueur des gènes par chromosome.
</p>
</div>
</div>
<div id="descriptive-parameters" class="section level3">
<h3>7. Descriptive parameters</h3>
<p>Calculate the parameters of central tendency (mean, median, mode) and dispersion (variance, standard deviation, inter-quarterly deviation)</p>
<ul>
<li>for the genes of chromosome III;</li>
<li>for all yeast genes.</li>
</ul>
<pre><code>[1] 194   1</code></pre>
<pre><code>[1] &quot;data.frame&quot;</code></pre>
<pre><code>[1] &quot;numeric&quot;</code></pre>
<pre><code>length1 length2 length3 length4 length5 length6 
    741    1845    1374     780     630     525 </code></pre>
<pre><code>[1] &quot;Chromosome III contains 194 CDS&quot;</code></pre>
<pre><code>[1] 1169.521</code></pre>
<p><strong>Ah ah! (skeptical tone)</strong> The R function <code>sd()</code> does <strong>not</strong> compute the standard deviation of the input numbers (<span class="math inline">\(s\)</span>), but the <strong>estimate of the standard deviaiton of the population</strong> (<span class="math inline">\(\hat{\sigma}\)</span>)</p>
<p>Display these parameters on the histogram of gene length, using the function <code>arrows()</code></p>
</div>
<div id="intervalle-de-confiance" class="section level3">
<h3>8. Intervalle de confiance</h3>
<p>From genes of chromosome III (considered as the sample available in 1992), calculate a confidence interval around the mean, and formulate the interpretation of this confidence interval. Then evaluate whether or not this confidence interval covered the average population (all genes in the yeast genome, which became available 4 years after chromosome III).</p>
<p><span class="math display">\[ \bar{x} \pm \frac{\hat{\sigma}}{\sqrt(n)} \cdot t_{1-\alpha/2}^{n-1}\]</span></p>
<pre><code>[1] -1.972332</code></pre>
<p>Draw a polygon of frequencies indicating the number of genes per class (class medium).</p>
<p><img src="figures/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="distribution-of-gene-length" class="section level3">
<h3>9. Distribution of gene length</h3>
<ul>
<li><p>From the result of <code>hist()</code>, retrieve an array (in a variable of type <code>data.frame</code>) indicating the absolute frequencies (<code>count</code>) according to the median class size (<code>mids</code>),</p></li>
<li><p>Add to this table a column indicating the relative frequency of each class of gene length.</p></li>
<li><p>Add columns to this table indicating the <strong>empirical distribution function</strong> gene lengths (number of genes of a size less than or equal to each observed <span class="math inline">\(x\)</span> value, and relative frequency of this number).</p>
<ul>
<li>basic function: <code>cumsum()</code></li>
<li>advanced function:<code>ecdf()</code></li>
</ul></li>
<li><p>by using the functions <code>plot()</code> and <code>lines()</code>, draw a graph representing the absolute frequency per class (medians of classes in <span class="math inline">\(X\)</span>, counts in <span class="math inline">\(Y\)</span>), and the empirical distribution function.</p>
<ul>
<li>suggestion: superposez les ??utilisez le type de lignes “h” pour les fréquences de classe, et “l” ou “s” pour la fonction de répartition.</li>
</ul></li>
</ul>
</div>
<div id="randomly-expected-distribution-for-gene-length" class="section level3">
<h3>10. Randomly expected distribution for gene length</h3>
<p>Based on the genome size (12.156.679 bp) and codon genomic frequencies defined below, calculate the random expected gene length distribution, and add it to the graph.</p>
<p>You can download the genomic frequencies of all polynucleotides here: <a href="http://jvanheld.github.io/stat1/data/Saccharomyces_cerevisiae/oligo_freq/3nt_genomic_Saccharomyces_cerevisiae-ovlp-1str.tab">3nt_genomic_Saccharomyces_cerevisiae-ovlp-1str.tab</a></p>
<p>Alternative: create a variable <code>freq.3nt</code> and manually assign the values for the 4? required polynucleotides from the table below.</p>
<table>
<thead>
<tr class="header">
<th align="left">sequence</th>
<th align="right">frequency</th>
<th align="right">occurrences</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AAA</td>
<td align="right">0.0394</td>
<td align="right">478708</td>
</tr>
<tr class="even">
<td align="left">ATG</td>
<td align="right">0.0183</td>
<td align="right">221902</td>
</tr>
<tr class="odd">
<td align="left">TAA</td>
<td align="right">0.0224</td>
<td align="right">272041</td>
</tr>
<tr class="even">
<td align="left">TAG</td>
<td align="right">0.0129</td>
<td align="right">156668</td>
</tr>
<tr class="odd">
<td align="left">TGA</td>
<td align="right">0.0201</td>
<td align="right">244627</td>
</tr>
</tbody>
</table>
</div>
<div id="before-finishing-keep-track-of-your-session" class="section level3">
<h3>11. Before finishing : keep track of your session</h3>
<p>Tractability is an essential issue in science. The function <strong><em>R</em></strong> <code>sessionInfo()</code> provides a summary of the conditions of a work session: version of R, operator system, libraries of functions used.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=fr_FR.UTF-8    
 [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=fr_FR.UTF-8   
 [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.24

loaded via a namespace (and not attached):
 [1] compiler_3.6.1  magrittr_1.5    tools_3.6.1     htmltools_0.3.6
 [5] yaml_2.2.0      Rcpp_1.0.2      stringi_1.4.3   rmarkdown_1.15 
 [9] highr_0.8       stringr_1.4.0   xfun_0.9        digest_0.6.20  
[13] evaluate_0.14  </code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
