<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2019-09-13" />

<title>Analyse d’une table d’annotations génomiques</title>

<script src="02_yeast_annotations_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="02_yeast_annotations_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="02_yeast_annotations_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="02_yeast_annotations_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="02_yeast_annotations_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="02_yeast_annotations_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="02_yeast_annotations_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="02_yeast_annotations_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="02_yeast_annotations_files/navigation-1.1/tabsets.js"></script>
<script src="02_yeast_annotations_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Analyse d’une table d’annotations génomiques</h1>
<h3 class="subtitle">Probabilités et statistique pour la biologie (STAT1)</h3>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">2019-09-13</h4>

</div>


<div id="but-de-ce-tp" class="section level2">
<h2>But de ce TP</h2>
<p>Durant ce TP, vous serez amenés à effectuer les tâches suivantes:</p>
<ol style="list-style-type: decimal">
<li>Manipuler une table de données génomique (les annotations du génome de la levure).</li>
<li>Sélectionner un sous-ensemble des données en filtrant les lignes sur base d’un critère déterminé (type d’annotation, chromosome).</li>
<li>Générer des graphiques pour représenter différents aspects liés à ces données.</li>
<li>Calculer les estimateurs de tendance centrale et dispersion.</li>
<li>Calculer un intervalle de confiance autour de la moyenne.</li>
</ol>
</div>
<div id="rendu" class="section level2">
<h2>Rendu</h2>
<p>A la fin du TP, vous déposerez deux fichiers sur Ametice.</p>
<ol style="list-style-type: decimal">
<li>Votre <strong>code R</strong>.</li>
<li>Un <strong>rapport synthétique</strong> qui inclura une présentation des principaux résultats (figures, statistiques descriptives) et votre interprétation.</li>
</ol>
<div id="attendus-pour-le-code" class="section level3">
<h3>Attendus pour le code</h3>
<ol style="list-style-type: decimal">
<li><p>Le code doit être <strong>lisible et compréhensible</strong>: donnez à vos variables des noms indiquant explicitement ce qu’elles contiennent.</p></li>
<li><p>Le code devra être <strong>correctement documenté</strong> (le symbole <code>#</code> en début ou en milieu de ligne indique que le reste de cette ligne est un commentaire).</p></li>
</ol>
<ul>
<li><p>avant chaque bloc de code, expliquer ce que vous comptez faire, à quoi sert ce bloc de code;</p></li>
<li><p>si c’est utile, ajoutez quelques mots de commentaires pour justifier l’approche choisi;</p></li>
<li><p>chaque fois que vous déinifissez une variable, ajoutez sur la même ligne un commentaire indiquant ce que cette variable représente.</p></li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li>Le code doit être <strong>tansportable</strong>: après l’avoir téléchargé, on doit pouvoir l’exécuter sur une autre machine. Je testerai systématiquement si les fichiers de code peuvent être exécutés sur ma machine. Evitez donc tout recours à des chemins absolus (nous indiquons ci-dessous comment définir des chemins relatifs par rapport à la racine de votre compte).</li>
</ol>
</div>
<div id="attendus-pour-le-rapport-dinterpretation" class="section level3">
<h3>Attendus pour le rapport d’interprétation</h3>
<p>Le rapport doit être synthétique (1 page de texte maximum + autant de figures et tables que vous le désirez).</p>
<p>Chaque question doit être exprimée explicitement avant de présenter les résultats qui y répondent et de fournir l’interprétation de ces résultats.</p>
<p>Chaque figure ou table doit être documentée par une légende permettant à un lecteur naïf de comprendre ce qu’elle représente. L’interprétation des résultats affichés sur une figure ou table se trouvera dans le texte principal (avec une référence au numéro de figure ou table).</p>
</div>
</div>
<div id="exemple-historique-genome-de-la-levure" class="section level2">
<h2>Exemple historique: génome de la levure</h2>
<ul>
<li>1992: publication du premier chromosome eucaryote complet, le 3ème chromosome de la levure.</li>
<li>1996: publication du génome complet.</li>
</ul>
<p>Sur base des gènes dU 3ème chromosome (échantillon) on peut estimer la taille moyenne d’un gène de levure.</p>
<p><strong>Questions: </strong></p>
<ol style="list-style-type: lower-alpha">
<li><p>La moyenne d’échantillon (chromosome III) permettait-elle de prédire la moyenne de la population (génome complet) ?</p>
<p>Pour répondre à cette quesiton, nous imaginerons que nous sommes revenus en 1992, et utiliserons l’ensemble des gènes du chromosome III (considérés ici comme un échantillon du génome) pour estimer la taille moyenne des gènes pour l’ensemble du génome (la “population” de gènes&quot;).</p></li>
<li><p>Cet échantillon peut-il être qualifié de “simple et indépendant” ?</p></li>
</ol>
</div>
<div id="analyse-de-la-longueur-des-genes-de-la-levure-du-boulanger" class="section level2">
<h2>Analyse de la longueur des gènes de la levure du boulanger</h2>
<div class="figure" style="text-align: center">
<img src="figures/cds_length_histo-1.png" alt="Distribution of cds lengths for Saccharomyces cerevisiae. " width="90%" />
<p class="caption">
Distribution of cds lengths for Saccharomyces cerevisiae.
</p>
</div>
</div>
<div id="tutoriel" class="section level2">
<h2>Tutoriel</h2>
<p>Avant de passer aux exercices, nous vous montrons ici quelques éléments de base concernant la lecture, la manipulation et l’écriture des tableaux de données avec R.</p>
<div id="le-chemin-de-la-maison-manuel" class="section level3">
<h3>Le chemin de la maison (manuel)</h3>
<p>Nous allons créer un dossier pour ce tuto, en partant de la racine de nore compte.</p>
<p>Première possibilité (<strong>rapide mais peu élégant</strong>): entrer (manuellement) le chemin de la racine de votre compte dans une variable</p>
<p><code>dir.home &lt;- /le/chemin/de/ma/maison</code></p>
<ul>
<li>Avantage: rapide et pratique</li>
<li>Désavantage: non portable, ne fonctionnera que sur votre ordinateur</li>
</ul>
</div>
<div id="le-chemin-de-la-maison-automatique" class="section level3">
<h3>Le chemin de la maison (automatique)</h3>
<p>Une solution plus générale: utiliser la commande <strong>R</strong> <code>Sys.getenv()</code>.</p>
<ul>
<li>Invoquée sans paramètre, cette commande liste toutes les variables d’environnement (votre configuration système).</li>
<li>On peut restreindre l’output à une variable d’environnement donnée, par exemple <code>Sys.getenv(&quot;HOME&quot;)</code> retourne le chemin de la racine de votre compte.</li>
</ul>
<p><strong>Note:</strong> écriture équivalente sous Linux: le symbole tilde <code>~</code> indique également le chemin de la rachine de votre compte.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">## Identify the home directory </span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">## by getting the environment variable HOME </span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">dir.home &lt;-<span class="st"> </span><span class="kw">Sys.getenv</span>(<span class="st">&quot;HOME&quot;</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">print</span>(dir.home)</a></code></pre></div>
<pre><code>[1] &quot;/Users/jvanheld&quot;</code></pre>
</div>
<div id="creation-dun-dossier-pour-le-tp" class="section level3">
<h3>Création d’un dossier pour le TP</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">## Define a variable containing the path of the results for this tutorial</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">dir.tuto &lt;-<span class="st"> </span><span class="kw">file.path</span>(dir.home, <span class="st">&quot;stat1&quot;</span>, <span class="st">&quot;TP2&quot;</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw">print</span>(dir.tuto)</a></code></pre></div>
<pre><code>[1] &quot;/Users/jvanheld/stat1/TP2&quot;</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">## Create the directory for this tutorial</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw">dir.create</span>(<span class="dt">path =</span> dir.tuto, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">## Go to the tutorial directory</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw">setwd</span>(dir.tuto)</a>
<a class="sourceLine" id="cb5-6" data-line-number="6"></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">## List the files already present in the folder (if any)</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="kw">list.files</span>()</a></code></pre></div>
<pre><code>[1] &quot;3nt_genomic_Saccharomyces_cerevisiae-ovlp-1str.tab&quot;
[2] &quot;chrom_sizes.tsv&quot;                                   
[3] &quot;Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz&quot;        </code></pre>
</div>
<div id="telechargement-du-fichier-gtf-a-partir-densemblgenomes" class="section level3">
<h3>Téléchargement du fichier GTF à partir d’EnsemblGenomes</h3>
<p><strong>Astuce: </strong> avant de télécharger le fichier d’annotations (GTF) depuis EnsemblGenomes vers notre ordinateur, nous allons vérifier s’il est déjà présent (et dans ce cas on ne le re-télécharge pas).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co">## Define the URL of the annotation file (GTF-formatted)</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">gtf.URL &lt;-<span class="st"> &quot;ftp://ftp.ensemblgenomes.org/pub/release-37/fungi/gtf/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz&quot;</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">## Define the path where the local copy will be stored</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">local.GTF &lt;-<span class="st"> </span><span class="kw">file.path</span>(dir.tuto, <span class="st">&quot;Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz&quot;</span>)</a>
<a class="sourceLine" id="cb7-6" data-line-number="6"></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="co">## If the local file file laready exists, skip the download</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="cf">if</span> (<span class="kw">file.exists</span>(local.GTF)) {</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">  <span class="kw">message</span>(<span class="st">&quot;GTF file already exists in the tutorial folder: &quot;</span>, local.GTF)</a>
<a class="sourceLine" id="cb7-10" data-line-number="10">} <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb7-11" data-line-number="11">  <span class="co">## Download annotation table in GTF format</span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12">  <span class="kw">download.file</span>(<span class="dt">url =</span> gtf.URL, <span class="dt">destfile =</span> local.GTF)</a>
<a class="sourceLine" id="cb7-13" data-line-number="13">}</a></code></pre></div>
</div>
<div id="chargement-dun-tableau-de-donnees" class="section level3">
<h3>Chargement d’un tableau de données</h3>
<p>R comporte plusieurs types de structures tabulaires (matrix, data.frame, table).</p>
<p>La structure la plus couramment utilisée est le <code>data.frame</code>, qui consiste en un tableau de valeurs (numériques ou chaînes de caractères) dont les lignes et les colonnes sont associées à des noms.</p>
<p>La fonction <code>read.table()</code> permet de lire un fichier texte contenant un tableau de données, et de stocker le contenu dans une variable.</p>
<p>Plusieurs fonctions dérivées de <code>read.table()</code> facilitent la lecture de différents types de formats:</p>
<ul>
<li><code>read.delim()</code> pour les fichiers dont les colonnes sont délimitées par un caractère particulier (généralement la tabulation, représentée par &quot;).</li>
<li><code>read.csv()</code> pour les fichiers “comma-searated values”.</li>
</ul>
<ol style="list-style-type: decimal">
<li>Téléchargez le fichier suivant sur votre ordinateur:</li>
</ol>
<ul>
<li><a href="../../data/Saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.37.gtf">Saccharomyces_cerevisiae.R64-1-1.37.gtf</a></li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>Chargez-le au moyen de la fonction read.table (pour cela vous devez remplacer le chemin ci-dessous par celui de votre ordinateur).</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co">## Read a GTF file with yeast genome annotations</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"></a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="co">## Load the feature table</span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4">feature.table &lt;-<span class="st"> </span><span class="kw">read.table</span>(</a>
<a class="sourceLine" id="cb8-5" data-line-number="5">  local.GTF, </a>
<a class="sourceLine" id="cb8-6" data-line-number="6">  <span class="dt">comment.char =</span> <span class="st">&quot;#&quot;</span>, </a>
<a class="sourceLine" id="cb8-7" data-line-number="7">  <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, </a>
<a class="sourceLine" id="cb8-8" data-line-number="8">  <span class="dt">header=</span><span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb8-9" data-line-number="9">  <span class="dt">row.names=</span><span class="ot">NULL</span>)</a>
<a class="sourceLine" id="cb8-10" data-line-number="10"></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co">## The bed format does not contain any column header, </span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="co">## so we set it manually based on the description of the format, </span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="co">## found here: </span></a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="co">##     http://www.ensembl.org/info/website/upload/gff.html</span></a>
<a class="sourceLine" id="cb8-15" data-line-number="15"><span class="kw">names</span>(feature.table) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;seqname&quot;</span>, <span class="st">&quot;source&quot;</span>, <span class="st">&quot;feature&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;end&quot;</span>, <span class="st">&quot;score&quot;</span>, <span class="st">&quot;strand&quot;</span>, <span class="st">&quot;frame&quot;</span>, <span class="st">&quot;attribute&quot;</span>)</a></code></pre></div>
</div>
<div id="exploration-du-contenu-dun-tableau-de-donnees" class="section level3">
<h3>Exploration du contenu d’un tableau de données</h3>
<p>La première chose à faire après avoir chargé un tableau de données est de vérifier ses dimensions</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw">dim</span>(feature.table) <span class="co">## Dimensions of the tbale</span></a></code></pre></div>
<pre><code>[1] 43028     9</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">nrow</span>(feature.table) <span class="co">## Number of rows</span></a></code></pre></div>
<pre><code>[1] 43028</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw">ncol</span>(feature.table) <span class="co">## Number of columns</span></a></code></pre></div>
<pre><code>[1] 9</code></pre>
<p>L’affichage du tableau d’annotations complet ne serait pas très lisible, puisqu’il comporte des dizaines de milliers de lignes.</p>
<p>Nous pouvons afficher les premières lignes avec la fonction <code>head()</code>.</p>
<p><strong>Note: </strong> la dernière colonne est particulièrement lourde (elle contient un tas d’information). Nous verrons plus loin comment sélectionner un sous-ensemble des colonnes pour alléger l’affichage.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="co">## Display the 5 first rows of the feature table</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="kw">head</span>(feature.table, <span class="dt">n =</span> <span class="dv">5</span>) </a></code></pre></div>
<pre><code>  seqname source     feature start  end score strand frame
1      IV    SGD        gene  1802 2953     .      +     .
2      IV    SGD  transcript  1802 2953     .      +     .
3      IV    SGD        exon  1802 2953     .      +     .
4      IV    SGD         CDS  1802 2950     .      +     0
5      IV    SGD start_codon  1802 1804     .      +     0
                                                                                                                                                                                                                                    attribute
1                                                                                                                                                              gene_id YDL248W; gene_name COS7; gene_source SGD; gene_biotype protein_coding;
2                                                       gene_id YDL248W; transcript_id YDL248W; gene_name COS7; gene_source SGD; gene_biotype protein_coding; transcript_name COS7; transcript_source SGD; transcript_biotype protein_coding;
3                     gene_id YDL248W; transcript_id YDL248W; exon_number 1; gene_name COS7; gene_source SGD; gene_biotype protein_coding; transcript_name COS7; transcript_source SGD; transcript_biotype protein_coding; exon_id YDL248W.1;
4 gene_id YDL248W; transcript_id YDL248W; exon_number 1; gene_name COS7; gene_source SGD; gene_biotype protein_coding; transcript_name COS7; transcript_source SGD; transcript_biotype protein_coding; protein_id YDL248W; protein_version 1;
5                                        gene_id YDL248W; transcript_id YDL248W; exon_number 1; gene_name COS7; gene_source SGD; gene_biotype protein_coding; transcript_name COS7; transcript_source SGD; transcript_biotype protein_coding;</code></pre>
<p>La fonction <code>tail()</code> affiche les dernières lignes:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co">## Display the 5 last rows of the feature table</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw">tail</span>(feature.table, <span class="dt">n =</span> <span class="dv">5</span>) </a></code></pre></div>
<pre><code>      seqname source     feature start   end score strand frame
43024    Mito    SGD  transcript 85554 85709     .      +     .
43025    Mito    SGD        exon 85554 85709     .      +     .
43026    Mito    SGD         CDS 85554 85706     .      +     0
43027    Mito    SGD start_codon 85554 85556     .      +     0
43028    Mito    SGD  stop_codon 85707 85709     .      +     0
                                                                                                                                                                                            attribute
43024                                                     gene_id Q0297; transcript_id Q0297; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding;
43025                     gene_id Q0297; transcript_id Q0297; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding; exon_id Q0297.1;
43026 gene_id Q0297; transcript_id Q0297; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding; protein_id Q0297; protein_version 1;
43027                                      gene_id Q0297; transcript_id Q0297; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding;
43028                                      gene_id Q0297; transcript_id Q0297; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding;</code></pre>
<p>If you are using the <strong>RStudio</strong> environment, you can display the table in a dynamic viewer pane with the function <code>View()</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="co">## In RStudio, display the table in a separate tab</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw">View</span>(feature.table) </a></code></pre></div>
</div>
<div id="selection-de-sous-ensembles-dun-tableau" class="section level3">
<h3>Sélection de sous-ensembles d’un tableau</h3>
<p>Sélection d’une ligne spécifiée par son indice.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">feature.table[<span class="dv">12</span>,]</a></code></pre></div>
<pre><code>   seqname source    feature start  end score strand frame
12      IV    SGD stop_codon  3834 3836     .      +     0
                                                                                                                                                            attribute
12 gene_id YDL247W-A; transcript_id YDL247W-A; exon_number 1; gene_source SGD; gene_biotype protein_coding; transcript_source SGD; transcript_biotype protein_coding;</code></pre>
<p>Sélection d’une colonne spécifiée par son indice (affichage des premières valeurs seulement.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw">head</span>(feature.table[,<span class="dv">3</span>])</a></code></pre></div>
<pre><code>[1] gene        transcript  exon        CDS         start_codon stop_codon 
Levels: CDS exon gene start_codon stop_codon transcript</code></pre>
<p>Selection d’une cellule en combinant indices de ligne et colonne.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">feature.table[<span class="dv">12</span>, <span class="dv">3</span>]</a></code></pre></div>
<pre><code>[1] stop_codon
Levels: CDS exon gene start_codon stop_codon transcript</code></pre>
<p>Sélection d’un bloc de colonnes et/ou de lignes.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">feature.table[<span class="dv">100</span><span class="op">:</span><span class="dv">105</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</a></code></pre></div>
<pre><code>    seqname source     feature start   end score
100      IV    SGD         CDS 34240 36477     .
101      IV    SGD start_codon 36475 36477     .
102      IV    SGD  stop_codon 34237 34239     .
103      IV    SGD        gene 36797 38173     .
104      IV    SGD  transcript 36797 38173     .
105      IV    SGD        exon 36797 38173     .</code></pre>
<p>Sélection de colonnes “à la carte” (ici, les coordonnées génomiques de chaque “feature”): chromosome, début, fin, brin.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">feature.table[<span class="dv">100</span><span class="op">:</span><span class="dv">105</span>, <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">7</span>)]</a></code></pre></div>
<pre><code>    seqname start   end strand
100      IV 34240 36477      -
101      IV 36475 36477      -
102      IV 34237 34239      -
103      IV 36797 38173      +
104      IV 36797 38173      +
105      IV 36797 38173      +</code></pre>
<p>Sélectionner une colonne sur base de son nom.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="co">## Select the &quot;start&quot; column and print the 100 first results</span></a>
<a class="sourceLine" id="cb30-2" data-line-number="2"><span class="kw">head</span>(feature.table<span class="op">$</span>start, <span class="dt">n=</span><span class="dv">100</span>)</a></code></pre></div>
<pre><code>  [1]  1802  1802  1802  1802  1802  2951  3762  3762  3762  3762  3762
 [12]  3834  5985  5985  5985  5985  5985  7812  8683  8683  8683  8686
 [23]  9754  8683 11657 11657 11657 11660 13358 11657 16204 16204 16204
 [34] 16204 16204 17224 17577 17577 17577 17580 18564 17577 18959 18959
 [45] 18959 18959 18959 19310 20635 20635 20635 20635 20635 21004 22471
 [56] 22471 22471 22474 22606 22471 22823 22823 22823 22823 22823 25874
 [67] 26403 26403 26403 26406 28773 26403 28985 28985 28985 28988 30452
 [78] 28985 30657 30657 30657 30657 30657 31827 32296 32296 32296 32296
 [89] 32296 33232 33415 33415 33415 33418 33916 33415 34237 34237 34237
[100] 34240</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="co">## Print the 20 first values of the &quot;feature&quot; field, which indicates the feature type</span></a>
<a class="sourceLine" id="cb32-2" data-line-number="2"><span class="kw">head</span>(feature.table<span class="op">$</span>feature, <span class="dt">n=</span><span class="dv">20</span>)</a></code></pre></div>
<pre><code> [1] gene        transcript  exon        CDS         start_codon
 [6] stop_codon  gene        transcript  exon        CDS        
[11] start_codon stop_codon  gene        transcript  exon       
[16] CDS         start_codon stop_codon  gene        transcript 
Levels: CDS exon gene start_codon stop_codon transcript</code></pre>
<p>Sélection de plusieurs colonnes sur base de leurs noms.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="co">## Select the &quot;start&quot; column and print the 100 first results</span></a>
<a class="sourceLine" id="cb34-2" data-line-number="2">feature.table[<span class="dv">100</span><span class="op">:</span><span class="dv">106</span>, <span class="kw">c</span>(<span class="st">&quot;seqname&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;end&quot;</span>, <span class="st">&quot;strand&quot;</span>)]</a></code></pre></div>
<pre><code>    seqname start   end strand
100      IV 34240 36477      -
101      IV 36475 36477      -
102      IV 34237 34239      -
103      IV 36797 38173      +
104      IV 36797 38173      +
105      IV 36797 38173      +
106      IV 36797 38170      +</code></pre>
<p><strong>Note</strong>: il est également possible de nommer les lignes d’un data.frame mais le tableau GTF ne se prête pas à cela. Nous verrons d’autres exemples ultérieurement.</p>
</div>
<div id="selection-dun-sous-ensemble-de-lignes-sur-base-du-contenu-dune-colonne" class="section level3">
<h3>Sélection d’un sous-ensemble de lignes sur base du contenu d’une colonne</h3>
<p>La fonction <code>subset()</code> permet de sélectionner un sous-ensemble des lignes d’un data.frame sur base d’une condition appliquée à une ou plusieurs colonnes.</p>
<p>Nous pouvons l’appliquer pour sélectionner le sous-ensemble des lignes du tableau d’annotations correspondant à des séquences codantes (CDS).</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="co">## Select subset of features having &quot;cds&quot; as &quot;feature&quot; attribute</span></a>
<a class="sourceLine" id="cb36-2" data-line-number="2">cds &lt;-<span class="st"> </span><span class="kw">subset</span>(feature.table, feature<span class="op">==</span><span class="st">&quot;cds&quot;</span>)</a>
<a class="sourceLine" id="cb36-3" data-line-number="3"></a>
<a class="sourceLine" id="cb36-4" data-line-number="4"><span class="kw">nrow</span>(feature.table) <span class="co">## Count the number of features</span></a></code></pre></div>
<pre><code>[1] 43028</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="kw">nrow</span>(cds) <span class="co">## Count the number of cds</span></a></code></pre></div>
<pre><code>[1] 0</code></pre>
</div>
<div id="decompte-par-valeur" class="section level3">
<h3>Décompte par valeur</h3>
<p>La fonction <code>table()</code> permet de compter le nombre d’occurrences de chaque valeur dans un vecteur ou un tableau. Quelques exemples d’utilisation ci-dessous.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="co">## Count the number of featues per chromosome</span></a>
<a class="sourceLine" id="cb40-2" data-line-number="2"><span class="kw">table</span>(feature.table<span class="op">$</span>seqname)</a></code></pre></div>
<pre><code>
   I   II  III   IV   IX Mito    V   VI  VII VIII    X   XI  XII XIII  XIV 
 759 2912 1210 5374 1567  327 2159  946 3856 2054 2617 2231 3789 3311 2774 
  XV  XVI 
3846 3296 </code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="co">## Count the number of features per type</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2"><span class="kw">table</span>(feature.table<span class="op">$</span>feature)</a></code></pre></div>
<pre><code>
        CDS        exon        gene start_codon  stop_codon  transcript 
       7050        7872        7445        6700        6516        7445 </code></pre>
<p>On peut calculer des tables de contingence en comptant le nombre de combinaisons entre 2 vecteurs (ou 2 colonnes d’un tableau).</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="co">##  Table with two vectors</span></a>
<a class="sourceLine" id="cb44-2" data-line-number="2"><span class="kw">table</span>(feature.table<span class="op">$</span>feature, feature.table<span class="op">$</span>seqname)</a></code></pre></div>
<pre><code>             
                I  II III  IV  IX Mito   V  VI VII VIII   X  XI XII XIII
  CDS         122 492 194 895 255   59 345 151 619  346 422 361 615  544
  exon        137 525 224 961 288   94 400 180 710  373 480 404 698  610
  gene        132 494 213 914 274   62 383 167 676  349 458 388 658  573
  start_codon 119 464 185 853 243   28 328 143 593  325 406 348 586  514
  stop_codon  117 443 181 837 233   22 320 138 582  312 393 342 574  497
  transcript  132 494 213 914 274   62 383 167 676  349 458 388 658  573
             
              XIV  XV XVI
  CDS         458 623 549
  exon        500 689 599
  gene        475 665 564
  start_codon 438 607 520
  stop_codon  428 597 500
  transcript  475 665 564</code></pre>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1"><span class="co">## Same result with a 2-column data frame</span></a>
<a class="sourceLine" id="cb46-2" data-line-number="2"><span class="kw">table</span>(feature.table[, <span class="kw">c</span>(<span class="st">&quot;feature&quot;</span>, <span class="st">&quot;seqname&quot;</span>)])</a></code></pre></div>
<pre><code>             seqname
feature         I  II III  IV  IX Mito   V  VI VII VIII   X  XI XII XIII
  CDS         122 492 194 895 255   59 345 151 619  346 422 361 615  544
  exon        137 525 224 961 288   94 400 180 710  373 480 404 698  610
  gene        132 494 213 914 274   62 383 167 676  349 458 388 658  573
  start_codon 119 464 185 853 243   28 328 143 593  325 406 348 586  514
  stop_codon  117 443 181 837 233   22 320 138 582  312 393 342 574  497
  transcript  132 494 213 914 274   62 383 167 676  349 458 388 658  573
             seqname
feature       XIV  XV XVI
  CDS         458 623 549
  exon        500 689 599
  gene        475 665 564
  start_codon 438 607 520
  stop_codon  428 597 500
  transcript  475 665 564</code></pre>
</div>
</div>
<div id="exercices" class="section level2">
<h2>Exercices</h2>
<div id="specifications-du-format-gtf" class="section level3">
<h3>1. Spécifications du format GTF</h3>
<p>Lisez les spécifications du format GTF.</p>
<ul>
<li>Ensembl (<a href="http://www.ensembl.org/info/website/upload/gff.html" class="uri">http://www.ensembl.org/info/website/upload/gff.html</a>)</li>
<li>UCSC (<a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format4" class="uri">https://genome.ucsc.edu/FAQ/FAQformat.html#format4</a>)</li>
</ul>
</div>
<div id="creation-dun-dossier-local-pour-le-tp" class="section level3">
<h3>2. Création d’un dossier local pour le TP</h3>
<p>Créez un dossier local (par exemple: <code>stat1/TP_levure</code> à partir de la racine de votre compte). Nous vous suggérons d’utiliser les fonctions suivantes:</p>
<ul>
<li><code>Sys.getenv(&quot;HOME&quot;)</code> (Linnux et Mac OS X), pour obtenir la racine de votre compte utilisateur;</li>
<li><code>file.path()</code> pour construire un chemin;</li>
<li><code>dir.create()</code> pour créer le dossier de ce TP. Lisez attendivement les options de cette fonction avec <code>help(dir.create)</code></li>
</ul>
</div>
<div id="localisation-du-fichier-dannotations" class="section level3">
<h3>3. Localisation du fichier d’annotations</h3>
<p>Localisez le fichier d’annotations du génome de la levure en format GTF dans ce dossier local.</p>
<ul>
<li>Site Ensembl Fungi: <a href="http://fungi.ensembl.org/" class="uri">http://fungi.ensembl.org/</a></li>
<li>Cliquez “Downloads” pour accéder au <a href="http://fungi.ensembl.org/info/website/ftp/">site ftp</a></li>
<li>Dans la bo^îte de recherche, tapez <em>“saccharomyces cerevisiae”</em> et suivez le lien “<a href="ftp://ftp.ensemblgenomes.org/pub/release-37/fungi/gtf/saccharomyces_cerevisiae">GTF</a>”</li>
<li>COpiez l’adresse (URL) du ichier <a href="ftp://ftp.ensemblgenomes.org/pub/release-37/fungi/gtf/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz">Saccharomyces_cerevisiae.R64-1-1.37.gtf.gz</a></li>
</ul>
</div>
<div id="telechargement-dun-fichier-a-partir-dun-site-ftp" class="section level3">
<h3>4. Téléchargement d’un fichier à partir d’un site ftp</h3>
<p>Fonctions suggérées:</p>
<ul>
<li><code>download.file()</code> (lisez l’aide pour conna^tre les arguments)</li>
</ul>
</div>
<div id="chargement-dune-table-de-donnees-en-r" class="section level3">
<h3>5. Chargement d’une table de données en R</h3>
<p>Ecrivez un script qui charge la table de données dans une variable nommée <code>feature.table</code>, en utilisant la fonction R <code>read.delim()</code>.</p>
<p>Veillez à ignorer les lignes de commentaires (qui commencent par un caractère <code>#</code>).</p>
</div>
<div id="calcul-de-la-longueur-des-genes-codants" class="section level3">
<h3>6. Calcul de la longueur des gènes codants</h3>
<ul>
<li>Ajoutez à la table d’annotations (<code>feature.table</code>) une colonne intitulée “length” qui indique la longueur de chaque élément génomique annoté.</li>
</ul>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1"><span class="co">## Add a colmn with feature lengths</span></a>
<a class="sourceLine" id="cb48-2" data-line-number="2">feature.table[, <span class="st">&quot;length&quot;</span>] &lt;-<span class="st"> </span>feature.table[, <span class="st">&quot;end&quot;</span>] <span class="op">-</span><span class="st"> </span>feature.table[, <span class="st">&quot;start&quot;</span>] <span class="op">+</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb48-3" data-line-number="3"></a>
<a class="sourceLine" id="cb48-4" data-line-number="4"><span class="co">## Add a colmn with feature lengths: equivalent result with simpler notation</span></a>
<a class="sourceLine" id="cb48-5" data-line-number="5">feature.table<span class="op">$</span>length &lt;-<span class="st"> </span>feature.table<span class="op">$</span>end <span class="op">-</span><span class="st"> </span>feature.table<span class="op">$</span>start <span class="op">+</span><span class="st"> </span><span class="dv">1</span></a></code></pre></div>
<ul>
<li><p>Comptez le nombre de lignes de la table correspondant à chaque type d’annotation (3ème colonne du GTF, “feature”).</p>
<ul>
<li>fonction <code>table()</code></li>
</ul></li>
</ul>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1"><span class="op">~</span><span class="kw">table</span>(feature.table<span class="op">$</span>feature)</a></code></pre></div>
<pre><code>~table(feature.table$feature)</code></pre>
<ul>
<li><p>Sélectionnez les lignes correspondant à des régions codantes (“CDS”)</p>
<ul>
<li>fonction <code>subset()</code></li>
</ul></li>
</ul>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1">cds &lt;-<span class="st"> </span><span class="kw">subset</span>(feature.table, feature<span class="op">==</span><span class="st">&quot;CDS&quot;</span>)</a></code></pre></div>
<ul>
<li><p>Comptez le nombre de CDS par chromosome.</p>
<ul>
<li>fonction <code>table()</code></li>
</ul></li>
</ul>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1"><span class="kw">table</span>(cds<span class="op">$</span>seqname)</a></code></pre></div>
<pre><code>
   I   II  III   IV   IX Mito    V   VI  VII VIII    X   XI  XII XIII  XIV 
 122  492  194  895  255   59  345  151  619  346  422  361  615  544  458 
  XV  XVI 
 623  549 </code></pre>
<ul>
<li>Chargez la table des tailles de chromosomes <a href="../../data/Saccharomyces_cerevisiae/chrom_sizes.tsv">chrom_sizes.tsv</a>, et calculez la densité de gènes pour chaque chromosome (nombre de gènes par Mb).</li>
</ul>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1"><span class="co">## Download tab-delimited file with chromosome sizes (unless already there)</span></a>
<a class="sourceLine" id="cb54-2" data-line-number="2">annot.url &lt;-<span class="st"> &quot;http://jvanheld.github.io/stat1/data/Saccharomyces_cerevisiae/chrom_sizes.tsv&quot;</span></a>
<a class="sourceLine" id="cb54-3" data-line-number="3">chrom.size.file &lt;-<span class="st"> </span><span class="kw">file.path</span>(dir.tuto, <span class="st">&quot;chrom_sizes.tsv&quot;</span>)</a>
<a class="sourceLine" id="cb54-4" data-line-number="4"></a>
<a class="sourceLine" id="cb54-5" data-line-number="5"><span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(chrom.size.file)) {</a>
<a class="sourceLine" id="cb54-6" data-line-number="6">    <span class="kw">download.file</span>(annot.url, <span class="dt">destfile =</span> chrom.size.file)</a>
<a class="sourceLine" id="cb54-7" data-line-number="7">}</a>
<a class="sourceLine" id="cb54-8" data-line-number="8"></a>
<a class="sourceLine" id="cb54-9" data-line-number="9"><span class="co">## Read chromosome sizes</span></a>
<a class="sourceLine" id="cb54-10" data-line-number="10">chrom.size &lt;-<span class="st"> </span><span class="kw">read.delim</span>(</a>
<a class="sourceLine" id="cb54-11" data-line-number="11">  <span class="dt">file =</span> chrom.size.file,</a>
<a class="sourceLine" id="cb54-12" data-line-number="12">  <span class="dt">header =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb54-13" data-line-number="13"></a>
<a class="sourceLine" id="cb54-14" data-line-number="14"><span class="co">## Assign a name to the columns</span></a>
<a class="sourceLine" id="cb54-15" data-line-number="15"><span class="kw">names</span>(chrom.size) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chromID&quot;</span>, <span class="st">&quot;size&quot;</span>)</a>
<a class="sourceLine" id="cb54-16" data-line-number="16"><span class="co"># View(chrom.size)</span></a>
<a class="sourceLine" id="cb54-17" data-line-number="17"></a>
<a class="sourceLine" id="cb54-18" data-line-number="18"><span class="co">## print the size of hte third chromosome</span></a>
<a class="sourceLine" id="cb54-19" data-line-number="19">chrom.size[<span class="st">&quot;III&quot;</span>, <span class="st">&quot;size&quot;</span>]</a></code></pre></div>
<pre><code>[1] 316617</code></pre>
</div>
<div id="histogramme-de-la-longueur-des-genes" class="section level3">
<h3>6. Histogramme de la longueur des gènes</h3>
<p>Au moyen de la fonction <code>hist()</code>, dessinez un histogramme représentant la distribution de longueur des CDS.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" data-line-number="1"><span class="kw">hist</span>(cds<span class="op">$</span>length)</a></code></pre></div>
<p><img src="figures/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Choisissez les intervalles de classe de façon à ce que l’histogramme soit informaatif (ni trop ni trop peu de classes).</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" data-line-number="1"><span class="co">## Take more or less 100 bins</span></a>
<a class="sourceLine" id="cb57-2" data-line-number="2">h &lt;-<span class="st"> </span><span class="kw">hist</span>(cds<span class="op">$</span>length, <span class="dt">breaks=</span><span class="dv">100</span>)</a></code></pre></div>
<p><img src="figures/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Récupérez le résultat de <code>hist()</code> dans une variable nommée <code>cds.length.hist</code>.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" data-line-number="1"><span class="co">## Define breaks exactly in the way you wish</span></a>
<a class="sourceLine" id="cb58-2" data-line-number="2">cds.length.hist &lt;-<span class="st"> </span><span class="kw">hist</span>(cds<span class="op">$</span>length, <span class="dt">breaks=</span><span class="kw">seq</span>(<span class="dt">from=</span><span class="dv">0</span>, <span class="dt">to=</span><span class="kw">max</span>(cds<span class="op">$</span>length)<span class="op">+</span><span class="dv">100</span>, <span class="dt">by=</span><span class="dv">100</span>))</a></code></pre></div>
<p><img src="figures/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Imprimez le résultat à l’écran (<code>print()</code>) et analysez la structure de la variable <code>cds.length.hist</code> (il s’agit d’une variable de type liste).</p>
<p>Fonctions utiles:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1"><span class="co">## Display the values used to draw the histogram</span></a>
<a class="sourceLine" id="cb59-2" data-line-number="2"><span class="kw">print</span>(cds.length.hist )</a></code></pre></div>
<pre><code>$breaks
  [1]     0   100   200   300   400   500   600   700   800   900  1000
 [12]  1100  1200  1300  1400  1500  1600  1700  1800  1900  2000  2100
 [23]  2200  2300  2400  2500  2600  2700  2800  2900  3000  3100  3200
 [34]  3300  3400  3500  3600  3700  3800  3900  4000  4100  4200  4300
 [45]  4400  4500  4600  4700  4800  4900  5000  5100  5200  5300  5400
 [56]  5500  5600  5700  5800  5900  6000  6100  6200  6300  6400  6500
 [67]  6600  6700  6800  6900  7000  7100  7200  7300  7400  7500  7600
 [78]  7700  7800  7900  8000  8100  8200  8300  8400  8500  8600  8700
 [89]  8800  8900  9000  9100  9200  9300  9400  9500  9600  9700  9800
[100]  9900 10000 10100 10200 10300 10400 10500 10600 10700 10800 10900
[111] 11000 11100 11200 11300 11400 11500 11600 11700 11800 11900 12000
[122] 12100 12200 12300 12400 12500 12600 12700 12800 12900 13000 13100
[133] 13200 13300 13400 13500 13600 13700 13800 13900 14000 14100 14200
[144] 14300 14400 14500 14600 14700 14800

$counts
  [1] 288 236 292 689 394 322 331 282 317 300 295 292 250 316 233 215 194
 [18] 190 170 128 131 109  99  81  81  72  80  51  53  39  39  38  45  40
 [35]  27  28  26  14  17  40  27  12  14  17  12  13  11  10   4  11   6
 [52]   3   3   7   1   4   8   3   2   1   3   2   0   2   3   3   4   0
 [69]   1   0   0   2   1   0   4   0   0   0   1   0   1   0   1   1   0
 [86]   0   0   0   1   0   0   0   2   0   1   0   0   0   1   0   0   0
[103]   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0
[120]   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0
[137]   0   0   0   0   0   0   0   0   0   0   0   1

$density
  [1] 4.085106e-04 3.347518e-04 4.141844e-04 9.773050e-04 5.588652e-04
  [6] 4.567376e-04 4.695035e-04 4.000000e-04 4.496454e-04 4.255319e-04
 [11] 4.184397e-04 4.141844e-04 3.546099e-04 4.482270e-04 3.304965e-04
 [16] 3.049645e-04 2.751773e-04 2.695035e-04 2.411348e-04 1.815603e-04
 [21] 1.858156e-04 1.546099e-04 1.404255e-04 1.148936e-04 1.148936e-04
 [26] 1.021277e-04 1.134752e-04 7.234043e-05 7.517730e-05 5.531915e-05
 [31] 5.531915e-05 5.390071e-05 6.382979e-05 5.673759e-05 3.829787e-05
 [36] 3.971631e-05 3.687943e-05 1.985816e-05 2.411348e-05 5.673759e-05
 [41] 3.829787e-05 1.702128e-05 1.985816e-05 2.411348e-05 1.702128e-05
 [46] 1.843972e-05 1.560284e-05 1.418440e-05 5.673759e-06 1.560284e-05
 [51] 8.510638e-06 4.255319e-06 4.255319e-06 9.929078e-06 1.418440e-06
 [56] 5.673759e-06 1.134752e-05 4.255319e-06 2.836879e-06 1.418440e-06
 [61] 4.255319e-06 2.836879e-06 0.000000e+00 2.836879e-06 4.255319e-06
 [66] 4.255319e-06 5.673759e-06 0.000000e+00 1.418440e-06 0.000000e+00
 [71] 0.000000e+00 2.836879e-06 1.418440e-06 0.000000e+00 5.673759e-06
 [76] 0.000000e+00 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00
 [81] 1.418440e-06 0.000000e+00 1.418440e-06 1.418440e-06 0.000000e+00
 [86] 0.000000e+00 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00
 [91] 0.000000e+00 0.000000e+00 2.836879e-06 0.000000e+00 1.418440e-06
 [96] 0.000000e+00 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00
[101] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[106] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[111] 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00 0.000000e+00
[116] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[121] 0.000000e+00 0.000000e+00 1.418440e-06 0.000000e+00 0.000000e+00
[126] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[131] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[136] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[141] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
[146] 0.000000e+00 0.000000e+00 1.418440e-06

$mids
  [1]    50   150   250   350   450   550   650   750   850   950  1050
 [12]  1150  1250  1350  1450  1550  1650  1750  1850  1950  2050  2150
 [23]  2250  2350  2450  2550  2650  2750  2850  2950  3050  3150  3250
 [34]  3350  3450  3550  3650  3750  3850  3950  4050  4150  4250  4350
 [45]  4450  4550  4650  4750  4850  4950  5050  5150  5250  5350  5450
 [56]  5550  5650  5750  5850  5950  6050  6150  6250  6350  6450  6550
 [67]  6650  6750  6850  6950  7050  7150  7250  7350  7450  7550  7650
 [78]  7750  7850  7950  8050  8150  8250  8350  8450  8550  8650  8750
 [89]  8850  8950  9050  9150  9250  9350  9450  9550  9650  9750  9850
[100]  9950 10050 10150 10250 10350 10450 10550 10650 10750 10850 10950
[111] 11050 11150 11250 11350 11450 11550 11650 11750 11850 11950 12050
[122] 12150 12250 12350 12450 12550 12650 12750 12850 12950 13050 13150
[133] 13250 13350 13450 13550 13650 13750 13850 13950 14050 14150 14250
[144] 14350 14450 14550 14650 14750

$xname
[1] &quot;cds$length&quot;

$equidist
[1] TRUE

attr(,&quot;class&quot;)
[1] &quot;histogram&quot;</code></pre>
<ul>
<li><code>class(cds.length.hist)</code></li>
<li><code>attributes(cds.length.hist)</code></li>
</ul>
<p>D’autres types de graphiques permettent d’explorer la distribution d’un ensemble des données. En particulier, les boîtes à moustaches (box plots) affichent, pour une série de données, la médiane, l’écart interquartile, un intervalle de confiance et les valeurs aberrantes.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" data-line-number="1"><span class="kw">boxplot</span>(length <span class="op">~</span><span class="st"> </span>seqname, <span class="dt">data =</span> cds, <span class="dt">col=</span><span class="st">&quot;palegreen&quot;</span>, <span class="dt">horizontal=</span><span class="ot">TRUE</span>, <span class="dt">las=</span><span class="dv">1</span>, <span class="dt">xlab=</span><span class="st">&quot;Gene length&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Chromosome&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/unnamed-chunk-24-1.png" alt="Boîte à moustache indiquant la distribution de longueur des gènes par chromosome. " width="672" />
<p class="caption">
Boîte à moustache indiquant la distribution de longueur des gènes par chromosome.
</p>
</div>
</div>
<div id="parametres-descriptifs" class="section level3">
<h3>7. Paramètres descriptifs</h3>
<p>Calculez les paramètres de tendance centrale (moyenne, médiane, mode) et de dispersion (variance, écart-type, écart inter-quartile)</p>
<ul>
<li>pour les gènes du chromosome III;</li>
<li>pour l’ensemble des gènes de la levure.</li>
</ul>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb62-1" data-line-number="1">x &lt;-<span class="st"> </span><span class="kw">subset</span>(cds, seqname <span class="op">==</span><span class="st"> &quot;III&quot;</span>, <span class="dt">select =</span> <span class="st">&quot;length&quot;</span>)</a>
<a class="sourceLine" id="cb62-2" data-line-number="2"><span class="kw">dim</span>(x) </a></code></pre></div>
<pre><code>[1] 194   1</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb64-1" data-line-number="1"><span class="kw">class</span>(x) <span class="co">## Ah ah, this is not a vector but a data.frame</span></a></code></pre></div>
<pre><code>[1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb66-1" data-line-number="1"><span class="co">## Convert the data frame into a vector</span></a>
<a class="sourceLine" id="cb66-2" data-line-number="2">x &lt;-<span class="st"> </span><span class="kw">unlist</span>(x)</a>
<a class="sourceLine" id="cb66-3" data-line-number="3"><span class="kw">class</span>(x)</a></code></pre></div>
<pre><code>[1] &quot;numeric&quot;</code></pre>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb68-1" data-line-number="1"><span class="kw">head</span>(x)</a></code></pre></div>
<pre><code>length1 length2 length3 length4 length5 length6 
    741    1845    1374     780     630     525 </code></pre>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb70-1" data-line-number="1"><span class="co">## Compute the mean, either manually or with the ad hoc R function</span></a>
<a class="sourceLine" id="cb70-2" data-line-number="2">n &lt;-<span class="st"> </span><span class="kw">length</span>(x) </a>
<a class="sourceLine" id="cb70-3" data-line-number="3"><span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;Chromosome III contains&quot;</span>, n, <span class="st">&quot;CDS&quot;</span>))</a></code></pre></div>
<pre><code>[1] &quot;Chromosome III contains 194 CDS&quot;</code></pre>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb72-1" data-line-number="1"><span class="kw">message</span>(<span class="st">&quot;Chromosome III contains &quot;</span>, n, <span class="st">&quot; CDS&quot;</span>)</a>
<a class="sourceLine" id="cb72-2" data-line-number="2"></a>
<a class="sourceLine" id="cb72-3" data-line-number="3">m &lt;-<span class="st"> </span><span class="kw">mean</span>(x)</a>
<a class="sourceLine" id="cb72-4" data-line-number="4"><span class="kw">print</span>(m)</a></code></pre></div>
<pre><code>[1] 1169.521</code></pre>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" data-line-number="1"><span class="kw">message</span>(<span class="st">&quot;mean(x) = &quot;</span>, <span class="kw">round</span>(m, <span class="dt">digits =</span> <span class="dv">1</span>))</a>
<a class="sourceLine" id="cb74-2" data-line-number="2"></a>
<a class="sourceLine" id="cb74-3" data-line-number="3"><span class="co">## Compute the mean manually to compare the result</span></a>
<a class="sourceLine" id="cb74-4" data-line-number="4">m.recalc &lt;-<span class="st"> </span><span class="kw">sum</span>(x)<span class="op">/</span>n</a>
<a class="sourceLine" id="cb74-5" data-line-number="5"><span class="kw">message</span>(<span class="st">&quot;Manually computed sample mean: &quot;</span>, <span class="kw">round</span>(<span class="dt">digits=</span><span class="dv">1</span>, m.recalc))</a>
<a class="sourceLine" id="cb74-6" data-line-number="6"></a>
<a class="sourceLine" id="cb74-7" data-line-number="7"><span class="co">## Compute manually standard dev of the sample</span></a>
<a class="sourceLine" id="cb74-8" data-line-number="8">sample.var &lt;-<span class="st"> </span><span class="kw">sum</span>((x <span class="op">-</span><span class="st"> </span>m)<span class="op">^</span><span class="dv">2</span>)<span class="op">/</span><span class="st"> </span>n</a>
<a class="sourceLine" id="cb74-9" data-line-number="9">sample.sd &lt;-<span class="st"> </span><span class="kw">sqrt</span>(sample.var)</a>
<a class="sourceLine" id="cb74-10" data-line-number="10"><span class="kw">message</span>(<span class="st">&quot;Sample standard dev =&quot;</span>, <span class="kw">round</span>(<span class="dt">digits=</span><span class="dv">1</span>, sample.sd))</a>
<a class="sourceLine" id="cb74-11" data-line-number="11"></a>
<a class="sourceLine" id="cb74-12" data-line-number="12"><span class="co">## Compute an estimate of the population standard deviation</span></a>
<a class="sourceLine" id="cb74-13" data-line-number="13">pop.sd.est &lt;-<span class="st"> </span><span class="kw">sqrt</span>(<span class="kw">sum</span>((x <span class="op">-</span><span class="st"> </span>m)<span class="op">^</span><span class="dv">2</span>) <span class="op">/</span><span class="st"> </span>(n<span class="dv">-1</span>))</a>
<a class="sourceLine" id="cb74-14" data-line-number="14"><span class="kw">message</span>(<span class="st">&quot;Sample-based estimate of population standard dev =&quot;</span>, <span class="kw">round</span>(<span class="dt">digits=</span><span class="dv">1</span>, pop.sd.est))</a>
<a class="sourceLine" id="cb74-15" data-line-number="15"></a>
<a class="sourceLine" id="cb74-16" data-line-number="16"><span class="co">## Compute the standard deviation with R function sd()</span></a>
<a class="sourceLine" id="cb74-17" data-line-number="17">R.sd &lt;-<span class="st"> </span><span class="kw">sd</span>(x)</a>
<a class="sourceLine" id="cb74-18" data-line-number="18"><span class="kw">message</span>(<span class="st">&quot;Result of R sd() function =&quot;</span>, <span class="kw">round</span>(<span class="dt">digits=</span><span class="dv">1</span>, R.sd))</a></code></pre></div>
<p><strong>Ah ah! (skeptical tone)</strong> The R function <code>sd()</code> does <strong>not</strong> compute the standard deviation of the input numbers (<span class="math inline">\(s\)</span>), but the <strong>estimate of the standard deviaiton of the population</strong> (<span class="math inline">\(\hat{\sigma}\)</span>)</p>
<p>Affichez ces paramètres sur l’histogramme de la longueur des gènes, en utilisant la fonction <code>arrows()</code></p>
</div>
<div id="intervalle-de-confiance" class="section level3">
<h3>8. Intervalle de confiance</h3>
<p>A partir des gènes du chromosome III (considérés comme l’échantillon disponible en 1992), calculez un intervalle de confiance autour de la moyenne, et formulez l’interprétation de cet intervalle de confiance. Evaluez ensuite si cet intervalle de confiance recouvrait ou non la moyenne de la population (tous les gènes du génome de la levure, qui devint disponible 4 ans après le chromosome III).</p>
<p><span class="math display">\[ \bar{x} \pm \frac{\hat{\sigma}}{\sqrt(n)} \cdot t_{1-\alpha/2}^{n-1}\]</span></p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb75-1" data-line-number="1"><span class="co">## Define alpha, the risk</span></a>
<a class="sourceLine" id="cb75-2" data-line-number="2">alpha &lt;-<span class="st"> </span><span class="fl">0.05</span></a>
<a class="sourceLine" id="cb75-3" data-line-number="3"></a>
<a class="sourceLine" id="cb75-4" data-line-number="4"><span class="co">## Let us get the critical value for the t distribution</span></a>
<a class="sourceLine" id="cb75-5" data-line-number="5"><span class="kw">help</span>(<span class="st">&quot;TDist&quot;</span>)</a>
<a class="sourceLine" id="cb75-6" data-line-number="6"></a>
<a class="sourceLine" id="cb75-7" data-line-number="7"><span class="co">## Which value corresponds to alpha/2 </span></a>
<a class="sourceLine" id="cb75-8" data-line-number="8"></a>
<a class="sourceLine" id="cb75-9" data-line-number="9"><span class="co">## Beware ! by default the qt() function return the lower tail</span></a>
<a class="sourceLine" id="cb75-10" data-line-number="10"><span class="kw">qt</span>(<span class="dt">p =</span> alpha<span class="op">/</span><span class="dv">2</span>, <span class="dt">df =</span>  n <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)</a></code></pre></div>
<pre><code>[1] -1.972332</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb77-1" data-line-number="1"><span class="co">## For confidence intervals we need a positive t value, we thus take the upper tail</span></a>
<a class="sourceLine" id="cb77-2" data-line-number="2">t.value &lt;-<span class="st"> </span><span class="kw">qt</span>(<span class="dt">p =</span> alpha<span class="op">/</span><span class="dv">2</span>, <span class="dt">df =</span>  n <span class="op">-</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">lower.tail =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb77-3" data-line-number="3"></a>
<a class="sourceLine" id="cb77-4" data-line-number="4">IC.min &lt;-<span class="st"> </span>m <span class="op">-</span><span class="st"> </span>pop.sd.est <span class="op">*</span><span class="st"> </span>t.value <span class="op">/</span><span class="kw">sqrt</span>(n)</a>
<a class="sourceLine" id="cb77-5" data-line-number="5">IC.max &lt;-<span class="st"> </span>m <span class="op">+</span><span class="st"> </span>pop.sd.est <span class="op">*</span><span class="st"> </span>t.value <span class="op">/</span><span class="kw">sqrt</span>(n)</a>
<a class="sourceLine" id="cb77-6" data-line-number="6"></a>
<a class="sourceLine" id="cb77-7" data-line-number="7"><span class="kw">message</span>(<span class="st">&quot;Confidence interval: [&quot;</span>, </a>
<a class="sourceLine" id="cb77-8" data-line-number="8">        <span class="kw">round</span>(<span class="dt">digits=</span><span class="dv">1</span>, IC.min), </a>
<a class="sourceLine" id="cb77-9" data-line-number="9">        <span class="st">&quot;, &quot;</span>, </a>
<a class="sourceLine" id="cb77-10" data-line-number="10">        <span class="kw">round</span>(<span class="dt">digits=</span><span class="dv">1</span>, IC.max), <span class="st">&quot;]&quot;</span>)</a></code></pre></div>
<p>Dessinez un polygnone des fréquences indiquant le nombre de gènes par classe (milieux de classe).</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" data-line-number="1"><span class="co">## Draw a frequency polygon</span></a>
<a class="sourceLine" id="cb78-2" data-line-number="2"><span class="kw">plot</span>(cds.length.hist <span class="op">$</span>mids, cds.length.hist <span class="op">$</span>counts, </a>
<a class="sourceLine" id="cb78-3" data-line-number="3">     <span class="dt">main=</span><span class="st">&quot;Frequency polygon&quot;</span>, </a>
<a class="sourceLine" id="cb78-4" data-line-number="4">     <span class="dt">xlab=</span><span class="st">&quot;Gene length&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Number of genes&quot;</span>, </a>
<a class="sourceLine" id="cb78-5" data-line-number="5">     <span class="dt">type=</span><span class="st">&quot;l&quot;</span>, <span class="dt">col=</span><span class="st">&quot;darkgreen&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">5000</span>))</a>
<a class="sourceLine" id="cb78-6" data-line-number="6"><span class="kw">grid</span>()</a>
<a class="sourceLine" id="cb78-7" data-line-number="7"></a>
<a class="sourceLine" id="cb78-8" data-line-number="8"><span class="kw">arrows</span>(<span class="dt">x0 =</span> IC.min, <span class="dt">y0 =</span> <span class="dv">100</span>, <span class="dt">x1 =</span> IC.max, <span class="dt">y1=</span><span class="dv">100</span>, <span class="dt">length =</span> <span class="dv">0</span>, <span class="dt">angle =</span> <span class="dv">00</span>, <span class="dt">col=</span><span class="st">&quot;violet&quot;</span>, <span class="dt">lwd=</span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb78-9" data-line-number="9"><span class="kw">abline</span>(<span class="dt">v =</span> m, <span class="dt">col=</span><span class="st">&quot;violet&quot;</span>)</a>
<a class="sourceLine" id="cb78-10" data-line-number="10"></a>
<a class="sourceLine" id="cb78-11" data-line-number="11">pop.mean &lt;-<span class="st"> </span><span class="kw">mean</span>(cds<span class="op">$</span>length)</a>
<a class="sourceLine" id="cb78-12" data-line-number="12"><span class="kw">abline</span>(<span class="dt">v =</span> pop.mean, <span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</a>
<a class="sourceLine" id="cb78-13" data-line-number="13"></a>
<a class="sourceLine" id="cb78-14" data-line-number="14"><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;genome&quot;</span>, <span class="st">&quot;chr3&quot;</span>), <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;violet&quot;</span>), <span class="dt">lwd=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="figures/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="distribution-de-la-longueur-des-genes" class="section level3">
<h3>9. Distribution de la longueur des gènes</h3>
<ul>
<li><p>A partir du résultt de <code>hist()</code>, récupérez un tableau (dans une variable de type <code>data.frame</code>) indiquant les fréquences absolues (<code>count</code>) en fonction de la taille médiane des classes (<code>mids</code>),</p></li>
<li><p>Ajoutez à ce tableau une colonne indiquant la fréquence relative de chaque classe de longueurs de gènes.</p></li>
<li><p>Ajoutez à ce tableau des colonnes indiquant la <strong>fonction de répartition empirique</strong> des longueurs de gènes (nombre de gènes d’une taille inférieure ou égale à chaque valeur <span class="math inline">\(x\)</span> observée, et fréquence relative de ce nombre).</p>
<ul>
<li>fonction de base: <code>cumsum()</code></li>
<li>fonction avancée;<code>ecdf()</code></li>
</ul></li>
<li><p>Au moyen des fonctions <code>plot()</code> et <code>lines()</code>, dessinez un graphe représentant la fréquence absolue par classe (médianes de classes en <span class="math inline">\(X\)</span>, comptages en <span class="math inline">\(Y\)</span>), et la fonction de répartition empirique.</p>
<ul>
<li>suggestion: superposez les utilisez le type de lignes “h” pour les fréquences de classe, et “l” ou “s” pour la fonction de répartition.</li>
</ul></li>
</ul>
</div>
<div id="distribution-attendue-au-hasard-pour-la-longueur-des-genes" class="section level3">
<h3>10. Distribution attendue au hasard pour la longueur des gènes</h3>
<p>Sur base de la taille du génome (12.156.679 bp) et des fréquences génomiques de codons définies ci-dessous, calculez la distribution de longueurs de gènes attendue au hasard, et ajoutez-là au graphique.</p>
<p>Vous pouvez télécharger les fréquences génomiques de tous les trinucléotides ici: <a href="http://jvanheld.github.io/stat1/data/Saccharomyces_cerevisiae/oligo_freq/3nt_genomic_Saccharomyces_cerevisiae-ovlp-1str.tab">3nt_genomic_Saccharomyces_cerevisiae-ovlp-1str.tab</a></p>
<p>Alternative: créez une variable <code>freq.3nt</code> et assignez-y manuellement les valeurs pour les 4 nucléotides nécessaires, à partir de la table ci-dessous.</p>
<table>
<thead>
<tr class="header">
<th align="left">sequence</th>
<th align="right">frequency</th>
<th align="right">occurrences</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AAA</td>
<td align="right">0.0394</td>
<td align="right">478708</td>
</tr>
<tr class="even">
<td align="left">ATG</td>
<td align="right">0.0183</td>
<td align="right">221902</td>
</tr>
<tr class="odd">
<td align="left">TAA</td>
<td align="right">0.0224</td>
<td align="right">272041</td>
</tr>
<tr class="even">
<td align="left">TAG</td>
<td align="right">0.0129</td>
<td align="right">156668</td>
</tr>
<tr class="odd">
<td align="left">TGA</td>
<td align="right">0.0201</td>
<td align="right">244627</td>
</tr>
</tbody>
</table>
</div>
<div id="avant-de-terminer-conservez-la-trace-de-votre-session" class="section level3">
<h3>11. Avant de terminer : conservez la trace de votre session</h3>
<p>La traçabilité constitue un enjeu essentiel en sciences. La fonction <strong><em>R</em></strong> <code>sessionInfo()</code> fournit un résumé des conditions d’une session de travail: version de R, système opérateur, bibliothèques de fonctions utilisées.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb79-1" data-line-number="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.23

loaded via a namespace (and not attached):
 [1] compiler_3.6.1  magrittr_1.5    tools_3.6.1     htmltools_0.3.6
 [5] yaml_2.2.0      Rcpp_1.0.2      stringi_1.4.3   rmarkdown_1.14 
 [9] highr_0.8       stringr_1.4.0   xfun_0.8        digest_0.6.20  
[13] evaluate_0.14  </code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
