<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2020-03-06" />

<title>Practical – Microarray analysis – Data loading and exploration</title>

<script src="DenBoer_2009_load-and-explore_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="DenBoer_2009_load-and-explore_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="DenBoer_2009_load-and-explore_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="DenBoer_2009_load-and-explore_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="DenBoer_2009_load-and-explore_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="DenBoer_2009_load-and-explore_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="DenBoer_2009_load-and-explore_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="DenBoer_2009_load-and-explore_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="DenBoer_2009_load-and-explore_files/navigation-1.1/tabsets.js"></script>
<script src="DenBoer_2009_load-and-explore_files/navigation-1.1/codefolding.js"></script>




<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Practical – Microarray analysis – Data loading and exploration</h1>
<h3 class="subtitle">(STAT2)</h3>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">2020-03-06</h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this practical, we will load a dataset that will be used as study case to apply different approaches of multivariate analysis:</p>
<ul>
<li>data exploration</li>
<li>multidimensional scaling</li>
<li>differential analysis</li>
<li>clustering (unsupervised classification)</li>
<li>supervised classification</li>
</ul>
</div>
<div id="study-case" class="section level2">
<h2>Study case</h2>
<p><strong>Reference:</strong> Den Boer ML <em>et al.</em> (2009). A subtype of childhood acute lymphoblastic leukaemia with poor treatment outcome: a genome-wide classification study. Lancet Oncol. 2009 10:125-34.</p>
<ul>
<li><strong>DOI</strong>: [<a href="http://doi.org/10.1016/S1470-2045(08)70339-5">doi: 10.1016/S1470-2045(08)70339-5</a>]</li>
<li><strong>Pubmed</strong>: [<a href="https://www.ncbi.nlm.nih.gov/pubmed/19138562">PMID 19138562</a>].</li>
<li><strong>Raw data</strong> available at Gene Expression Omnibus, series [<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE13425">GSE13425</a>]</li>
<li><strong>Preprocessed data</strong>: <a href="https://github.com/jvanheld/stat1/data/DenBoer_2009" class="uri">https://github.com/jvanheld/stat1/data/DenBoer_2009</a></li>
</ul>
</div>
<div id="data-pre-processing" class="section level2">
<h2>Data pre-processing</h2>
<p>The raw microarray data has been pre-processed in order to dispose of a ready-to-use dataset. pre-processing included</p>
<ul>
<li>filtering of barely detected or poorly expressed genes,</li>
<li>log2 transformation to normalise the raw measurements</li>
<li>between-sample standardisation to enable comparison between the different samples.</li>
</ul>
</div>
<div id="availability-of-the-pre-processed-data" class="section level2">
<h2>Availability of the pre-processed data</h2>
<p>The preprocessed data is available here: <a href="https://github.com/jvanheld/stat1/data/DenBoer_2009" class="uri">https://github.com/jvanheld/stat1/data/DenBoer_2009</a>.</p>
<p>It contains the following files.</p>
<table>
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead>
<tr class="header">
<th>File</th>
<th>Contents</th>
<th>Structure</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="../../data/DenBoer_2009/GSE13425_group_descriptions.tsv.gz">GSE13425_group_descriptions.tsv.gz</a></td>
<td>Description of the patient groups</td>
<td>Tab-delimited file with one row per group and one column per type of description (group name, label)</td>
</tr>
<tr class="even">
<td><a href="../../data/DenBoer_2009/phenoData_GSE13425.tsv.gz">phenoData_GSE13425.tsv.gz</a></td>
<td>Metadata (sample descriptions)</td>
<td>Tab-delimited file with one row per biological sample (one per patient) and one column per type of information about the biological sample</td>
</tr>
<tr class="odd">
<td><a href="../../data/DenBoer_2009/GSE13425_AMP_Whole.tsv.gz">GSE13425_AMP_Whole.tsv.gz</a></td>
<td>Normalised microarray data</td>
<td>Tab-delimited file with one row per gene and one column per patient</td>
</tr>
<tr class="even">
<td><a href="../../data/DenBoer_2009/GSE13425_AMP_Whole.tsv.gz">GSE13425_AMP_Whole.tsv.gz</a></td>
<td>Detection status of each gene in each sample (Absent, Marginal, Present)</td>
<td>Tab-delimited file with one row per gene and one column per patient</td>
</tr>
</tbody>
</table>
</div>
<div id="data-download" class="section level2">
<h2>Data download</h2>
<p>Write an R script that perform the following operations</p>
<ul>
<li>creates a directory to store a local copy for this practical on your computer (<code>~/STAT2_CMB_practicals/den-boer-2009/</code>).</li>
<li>creates a sub-directory for the data (<code>~/STAT2_CMB_practicals/den-boer-2009/data/</code>).</li>
<li>lists the files available on the data Web site (<a href="https://github.com/jvanheld/stat1/data/DenBoer_2009" class="uri">https://github.com/jvanheld/stat1/data/DenBoer_2009</a>)</li>
<li>For each of these files, check if it is already present in your local local data directory, and download it if it is not already there.</li>
</ul>
</div>
<div id="data-loading" class="section level2">
<h2>Data loading</h2>
<p>Write an R script that loads the data tables from your local data directory.</p>
</div>
<div id="marginal-statistics" class="section level2">
<h2>Marginal statistics</h2>
<p>Write an R script that computes marginal statistics (mean, sd, min, percentiles 5, 10, 25, 50, 75, 90, 95, max, IQR) on each row of the normalised data table (one row corresponds to one gene).</p>
<p>Do the same for each column (patient).</p>
</div>
<div id="empirical-distributions" class="section level2">
<h2>Empirical distributions</h2>
<p>Draw a plot that displays the empirical distribution of normalised expression values in the whole data table.</p>
<p>Draw another plot that displays the empirical distribution of normalised expression values in the different samples (one polygon frequency per sample).</p>
</div>
<div id="sample-classes" class="section level2">
<h2>Sample classes</h2>
<p>Load the pheno table. We will use the three following columns:</p>
<ul>
<li><code>Sample.GEO.ID</code>: the identifier of the sample in the Gene Expression Omnibus (GEO) database.</li>
<li><code>Sample.title</code>: cancer type of each sample.</li>
<li><code>sample.labels</code>: A short label (1 or 2 letters) for each cancer type</li>
</ul>
<p>Cound the number of samples per group and draw a barplot with the result.</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
