<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2020-04-03" />

<title>Practical – Supervised classification – Microarrays</title>

<script src="DenBoer_2009_supervised-classification_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="DenBoer_2009_supervised-classification_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="DenBoer_2009_supervised-classification_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="DenBoer_2009_supervised-classification_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="DenBoer_2009_supervised-classification_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="DenBoer_2009_supervised-classification_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="DenBoer_2009_supervised-classification_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="DenBoer_2009_supervised-classification_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="DenBoer_2009_supervised-classification_files/navigation-1.1/tabsets.js"></script>
<script src="DenBoer_2009_supervised-classification_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Practical – Supervised classification – Microarrays</h1>
<h3 class="subtitle">STAT2</h3>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">2020-04-03</h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this practical we will define a pipeline to run supervised classification on a multivariate dataset (transcriptome microarrays of leukemia from Den Boer 2009).</p>
</div>
<div id="study-case" class="section level2">
<h2>Study case</h2>
<p><strong>Reference:</strong> Den Boer ML <em>et al.</em> (2009). A subtype of childhood acute lymphoblastic leukaemia with poor treatment outcome: a genome-wide classification study. Lancet Oncol. 2009 10:125-34.</p>
<ul>
<li><strong>DOI</strong>: [<a href="http://doi.org/10.1016/S1470-2045(08)70339-5">doi: 10.1016/S1470-2045(08)70339-5</a>]</li>
<li><strong>Pubmed</strong>: [<a href="https://www.ncbi.nlm.nih.gov/pubmed/19138562">PMID 19138562</a>].</li>
<li><strong>Raw data</strong> available at Gene Expression Omnibus, series [<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE13425">GSE13425</a>]</li>
<li><strong>Preprocessed data</strong>: <a href="https://github.com/jvanheld/stat1/tree/master/data/DenBoer_2009" class="uri">https://github.com/jvanheld/stat1/tree/master/data/DenBoer_2009</a>.</li>
</ul>
</div>
<div id="data-pre-processing" class="section level2">
<h2>Data pre-processing</h2>
<p>The raw microarray data has been pre-processed in order to dispose of a ready-to-use dataset. pre-processing included</p>
<ul>
<li>filtering of barely detected or poorly expressed genes,</li>
<li>log2 transformation to normalise the raw measurements</li>
<li>between-sample standardisation to enable comparison between the different samples.</li>
</ul>
</div>
<div id="defining-the-local-folders-for-this-practical" class="section level2">
<h2>Defining the local folders for this practical</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">####  Define Local directories and files ####</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">## Find the home directory</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">myHome &lt;-<span class="st"> </span><span class="kw">Sys.getenv</span>(<span class="st">&quot;HOME&quot;</span>)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">## Define the main directory for the  data and results</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">mainDir &lt;-<span class="st"> </span><span class="kw">file.path</span>(myHome,</a>
<a class="sourceLine" id="cb1-8" data-line-number="8">                     <span class="st">&quot;STAT2_CMB_practicals&quot;</span>,</a>
<a class="sourceLine" id="cb1-9" data-line-number="9">                     <span class="st">&quot;den-boer-2009&quot;</span>)</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="kw">dir.create</span>(<span class="dt">path =</span> mainDir, </a>
<a class="sourceLine" id="cb1-11" data-line-number="11">           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb1-12" data-line-number="12">           <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="kw">message</span>(<span class="st">&quot;Main dir: &quot;</span>, mainDir)</a>
<a class="sourceLine" id="cb1-14" data-line-number="14"></a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="co">## Define the dir where we will download the data</span></a>
<a class="sourceLine" id="cb1-16" data-line-number="16">destDir &lt;-<span class="st"> </span><span class="kw">file.path</span>(mainDir, <span class="st">&quot;data&quot;</span>)</a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="kw">dir.create</span>(<span class="dt">path =</span> destDir, </a>
<a class="sourceLine" id="cb1-18" data-line-number="18">           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb1-19" data-line-number="19">           <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-20" data-line-number="20"><span class="kw">message</span>(<span class="st">&quot;Local data dir: &quot;</span>, destDir)</a>
<a class="sourceLine" id="cb1-21" data-line-number="21"></a>
<a class="sourceLine" id="cb1-22" data-line-number="22"><span class="co">## Define a file where we wills tore the memory image</span></a>
<a class="sourceLine" id="cb1-23" data-line-number="23">memImageFile &lt;-<span class="st"> </span><span class="kw">file.path</span>(mainDir, <span class="st">&quot;DenBoerData_loaded.Rdata&quot;</span>)</a>
<a class="sourceLine" id="cb1-24" data-line-number="24"><span class="kw">message</span>(<span class="st">&quot;Memory image file of the loaded data: &quot;</span>, memImageFile)</a></code></pre></div>
</div>
<div id="re-loading-the-pre-proocessed-data" class="section level2">
<h2>Re-loading the pre-proocessed data</h2>
<p>The detailed procedure to build the data set from tab-delimited files was described in the practical <a href="DenBoer_2009_load-and-explore_solutions.html"><strong>Data loading and exploration</strong></a>.</p>
<p>At the end of this practical we saved a memory image with the data. We provide hereby a few lines of code that should enable you to reload this data.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">#### Load the data from a memory image ####</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Loading data from memory image file&quot;</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw">system.time</span>(<span class="kw">load</span>(memImageFile))</a></code></pre></div>
<pre><code>   user  system elapsed 
  0.653   0.046   0.705 </code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">message</span>(<span class="st">&quot;Data loaded&quot;</span>)</a></code></pre></div>
</div>
<div id="recommendations" class="section level2">
<h2>Recommendations</h2>
<ol start="0" style="list-style-type: decimal">
<li>Remember that we might need to run the same analyses</li>
</ol>
<ul>
<li>with different datasets</li>
<li>with different classifiers (to compare their performances)</li>
<li>with different parameter settings (to optimise a classifer)</li>
</ul>
<p>The code needs to be conceived with this in mind.</p>
<ol style="list-style-type: decimal">
<li><p>Split your code in modular sections. Whenever useful, write functions that can be called to run the different steps of an analysis.</p></li>
<li><p>The code should be re-usable by other people. For this, you need to document it.</p>
<ul>
<li>for developers who would reuse it and modify it (code documentation with hash comments)</li>
<li>for users (roxygen2 documentation for all the functions, tutorials, vignettes)</li>
</ul></li>
<li><p>Use the <code>message()</code> function (but not too many) to indicate when you start a task and finish a task + the main information (e.g. datadir, result dir, …).</p></li>
<li><p>Beware: traditionally, omics datasets (transcriptome, metabolome, proteome) are provided as matrices with 1 row per individual and 1 column per feature (variable),n but supervised classification methods expect martrices with individuals as rows and features as columns –&gt; you will need to transpose the matrix.</p></li>
</ol>
<div id="r-markdown-recommendations" class="section level3">
<h3>R markdown recommendations</h3>
<ol style="list-style-type: decimal">
<li>Name each chunk of code (you need to give a unique name)</li>
</ol>
</div>
</div>
<div id="pipeline-for-the-final-report" class="section level2">
<h2>Pipeline for the final report</h2>
<ol style="list-style-type: decimal">
<li>Load the data (done above)</li>
<li><p>Exploring the data (done last time)</p>
<ul>
<li>histogram with the distribution of values</li>
<li>boxplots (if we have too many individuals, take subsets of them)</li>
<li>PC plots (PC1 vs PC2, PC3 vs PC4, …) with a coloring of the classes</li>
</ul></li>
<li><p>Normalisation (no need here, the data we loaded was previously pre-processed)</p></li>
<li><p>Class filtering: suppress the classes with less than 30 individuals, because it will be very difficult to train a program properly with so scsarse data.</p></li>
<li><p>Supervised classification, with 3 steps</p>
<ul>
<li><p>Split the data into training and testing subsets</p>
<ul>
<li>stratified subsampling: same proprotions of the 4 classes in the training as in the testing</li>
<li>iterative subsampling with 2/3 for training and 1/3 for testing.</li>
<li>run the subsampling 10 times independently</li>
</ul></li>
<li>Train a classifier with the training subset</li>
<li><p>Evaluate the performances of the trained classifier with the testing subset 4.4. Apply the classification to an independent dataset</p></li>
</ul></li>
<li><p>Reduction of dimensionality. Choose one among the following approaches</p>
<ul>
<li>Very simple (rudimentary) criterion: variable ordering by decreasing variance, and keep the <span class="math inline">\(n\)</span> top variables. Analyse the impact of <span class="math inline">\(n\)</span></li>
<li>PCA: keep the <span class="math inline">\(n\)</span> first components</li>
<li>Differential analysis: keep the <span class="math inline">\(n\)</span> most significant genes from differential analysis (I will provide you with the sorted list of genes)</li>
</ul></li>
<li><p>Evaluate the impact of parameters on the performances. Examples</p>
<ul>
<li>KNN: impact of K, the number of neighbours</li>
<li>SVM: impact of the kernel</li>
<li>Random Forest: impact of number of iterations, … other parameters</li>
<li>Discriminant analysis: impact of the homoscedasticity assumption (compare LDA with QDA).</li>
</ul></li>
<li><p>Compare the performances of 2 different classification methods (e.g. LDA, QDA, SVM, RF, KNN)</p></li>
<li><p>Interpret the results</p>
<ul>
<li><p>statistical performances</p></li>
<li><p>try to explain why some method or some parameter combination works better for your dataset that some other one, relationship with the data structure, the dimensionality of the feature space, …</p></li>
<li><p>biological relevance</p></li>
</ul></li>
</ol>
</div>
<div id="hints" class="section level2">
<h2>Hints</h2>
<div id="relevant-packages-for-supervised-classification" class="section level3">
<h3>Relevant packages for supervised classification</h3>
<table>
<thead>
<tr class="header">
<th>Method</th>
<th>Package::Function(s)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Support vector machines (SVM)</td>
<td><code>e1071::svm()</code></td>
</tr>
<tr class="even">
<td>K nearest neighbours (KNN)</td>
<td><code>class::knn()</code></td>
</tr>
<tr class="odd">
<td>Random forest (RF)</td>
<td><code>randomForest::randomForest()</code></td>
</tr>
<tr class="even">
<td>Linear discriminant analysis</td>
<td><code>MASS::lda()</code></td>
</tr>
<tr class="odd">
<td>Utilities for supervised classification</td>
<td><code>caret</code> package</td>
</tr>
</tbody>
</table>
</div>
<div id="support" class="section level3">
<h3>Support</h3>
<ul>
<li>We interact via the Ametice forum for this job.</li>
<li>A support session by videoconference is scheduled on April 3, 10:00 - 12:00.</li>
</ul>
</div>
</div>
<div id="to-do" class="section level2">
<h2>To do</h2>
<ul>
<li><p>JvH: ask accounts for each student + a shared folder for this course, in order to give you access to the IFB-core RStudio server</p>
<p>done</p></li>
<li><p>JvH will add an independent dataset for evaluation, in the form of a Rdata memory image</p></li>
<li><p>JvH will give some hints, clues about some pieces of this work</p>
<ul>
<li><p>for each step indicate the relevant R functions (for example <code>sample()</code>, …)</p></li>
<li><p>for some step, provide an example of a properly documented function that does the job.</p></li>
</ul></li>
<li><p>Define some shared space for the code</p></li>
<li></li>
</ul>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
