---
title: "Practical -- Supervised classification -- Microarrays"
author: "Jacques van Helden"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    highlight: zenburn
    self_contained: no
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  ioslides_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 6
    fig_width: 7
    fonttheme: structurebold
    highlight: tango
    smaller: yes
    toc: yes
    widescreen: yes
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 6
    fig_width: 7
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
  word_document:
    toc: yes
    toc_depth: '3'
  slidy_presentation:
    fig_caption: yes
    fig_height: 6
    fig_width: 7
    highlight: tango
    incremental: no
    keep_md: yes
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
font-import: http://fonts.googleapis.com/css?family=Risque
subtitle:  STAT2
font-family: Garamond
transition: linear
---

```{r knitr_settings, include=FALSE, echo=FALSE, eval=TRUE}
library(knitr)
options(width = 300)
knitr::opts_chunk$set(
  fig.width = 7, fig.height = 5, 
  fig.path = 'figures/02_combinatorix_',
  fig.align = "center", 
  size = "tiny", 
  echo = TRUE, eval = TRUE, 
  warning = FALSE, message = FALSE, 
  results = TRUE, comment = "")
# knitr::asis_output("\\footnotesize")

```


## Introduction

In  this practical we will define a pipeline to run supervised classification on a multivariate dataset (transcriptome microarrays of leukemia from Den Boer 2009). 




## Study case

**Reference:** Den Boer ML *et al.* (2009). A subtype of childhood acute lymphoblastic leukaemia with poor treatment outcome: a genome-wide classification study. Lancet Oncol. 2009 10:125-34. 

- **DOI**: [[doi:Â 10.1016/S1470-2045(08)70339-5](http://doi.org/10.1016/S1470-2045(08)70339-5)]
- **Pubmed**: [[PMID 19138562](https://www.ncbi.nlm.nih.gov/pubmed/19138562)]. 
- **Raw data** available at Gene Expression Omnibus, series [[GSE13425](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE13425)]
- **Preprocessed data**: <https://github.com/jvanheld/stat1/tree/master/data/DenBoer_2009>.

## Data pre-processing

The raw microarray data has been pre-processed in order to dispose of a ready-to-use dataset. pre-processing included 

- filtering of barely detected or poorly expressed genes, 
- log2 transformation to normalise the raw measurements
- between-sample standardisation to enable comparison between the different samples.

## Defining the local folders for this practical

```{r folder_definition}
####  Define Local directories and files ####

## Find the home directory
myHome <- Sys.getenv("HOME")

## Define the main directory for the  data and results
mainDir <- file.path(myHome,
                     "STAT2_CMB_practicals",
                     "den-boer-2009")
dir.create(path = mainDir, 
           recursive = TRUE, 
           showWarnings = FALSE)
message("Main dir: ", mainDir)

## Define the dir where we will download the data
destDir <- file.path(mainDir, "data")
dir.create(path = destDir, 
           recursive = TRUE, 
           showWarnings = FALSE)
message("Local data dir: ", destDir)

## Define a file where we wills tore the memory image
memImageFile <- file.path(mainDir, "DenBoerData_loaded.Rdata")
message("Memory image file of the loaded data: ", memImageFile)

```


## Re-loading the pre-proocessed data

The detailed procedure to build the data set from tab-delimited files was described in the practical [**Data loading and exploration**](DenBoer_2009_load-and-explore_solutions.html).

At the end of this practical we saved a memory image with the data. We provide hereby a few lines of code that should enable you to reload this data. 

```{r load_data_from_memory_image}
#### Load the data from a memory image ####
message("Loading data from memory image file")
system.time(load(memImageFile))
message("Data loaded")
```

## Recommendations

0. Remember that we might need to run the same analyses

- with different datasets
- with different classifiers (to compare their performances)
- with different parameter settings (to optimise a classifer)


The code needs to be conceived with this in mind.

1. Split your code in modular sections. Whenever useful, write functions that can be called to run the different steps of an analysis. 

2. The code should be re-usable by other people. For this, you need to document it. 

    - for developers who would reuse it and modify it (code documentation with hash comments)
    - for users (roxygen2 documentation for all the functions, tutorials, vignettes)
    
3. Use the `message()` function (but not too many) to indicate when you start a task and finish a task + the main information (e.g. datadir, result dir, ...). 

4. Beware: traditionally, omics datasets (transcriptome, metabolome, proteome) are provided as matrices with 1 row per individual and 1 column per feature (variable),n but supervised classification methods expect martrices with individuals as rows and features as columns --> you will need to transpose the matrix. 


### R markdown recommendations

1. Name each chunk of code (you need to give a unique name)

## Pipeline for the final report

1. Load the data (done above)
2. Exploring the data (done last time)

    - histogram with the distribution of values
    - boxplots (if we have too many individuals, take subsets of them)
    - PC plots (PC1 vs PC2, PC3 vs PC4, ...) with a coloring of the classes
    
3. Normalisation (no need here, the data we loaded was previously pre-processed)

4. Supervised classification, with 3 steps

    4.1. Split the data into training and testing subsets
    4.2. Train a classifier with the training subset
    4.3. Evaluate the performances of the trained classifier with the testing subset
    4.4. Apply the classification to an independent dataset

5. Evaluate the impact of parameters on the performances. Examples

    - KNN: impact of K, the number of neighbours
    - SVM: impact of the kernel
    - Random Forest: impact of number of iterations, ... other parameters
    - Discriminant analysis: impact of the homoscedasticity assumption (compare LDA with QDA). 

6. Reduction of dimensionality. Choose one among the following approaches

    - Very simple (rudimentary) criterion: variable ordering by decreasing variance, and keep the $n$ top variables. Analyse the impact of $n$
    - PCA: keep the $n$ first components
    - Differential analysis: keep the $n$ most significant genes from differential analysis (I will provide you with the sorted list of genes)

7. Compare the performances of 2 different classification methods (e.g. LDA, QDA, SVM, RF, KNN)







